<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 14 Visualization | Orchestrating Microbiome Analysis with Bioconductor</title>
  <meta name="description" content="Chapter 14 Visualization | Orchestrating Microbiome Analysis with Bioconductor" />
  <meta name="generator" content="bookdown 0.36 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 14 Visualization | Orchestrating Microbiome Analysis with Bioconductor" />
  <meta property="og:type" content="book" />
  
  
  <meta name="github-repo" content="microbiome/OMA" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 14 Visualization | Orchestrating Microbiome Analysis with Bioconductor" />
  
  
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="multi-assay-analyses.html"/>
<link rel="next" href="training.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>
<script src="libs/kePrint-0.0.1/kePrint.js"></script>
<link href="libs/lightable-0.0.1/lightable.css" rel="stylesheet" />


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>
<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Orchestrating Microbiome Analysis with Bioconductor</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a></li>
<li class="part"><span><b>I Introduction</b></span></li>
<li class="chapter" data-level="1" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>1</b> Introduction</a></li>
<li class="chapter" data-level="2" data-path="packages.html"><a href="packages.html"><i class="fa fa-check"></i><b>2</b> Packages</a>
<ul>
<li class="chapter" data-level="2.1" data-path="packages.html"><a href="packages.html#package-installation"><i class="fa fa-check"></i><b>2.1</b> Package installation</a>
<ul>
<li class="chapter" data-level="2.1.1" data-path="packages.html"><a href="packages.html#packages_specific"><i class="fa fa-check"></i><b>2.1.1</b> Installing specific packages</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="packages.html"><a href="packages.html#ecosystem"><i class="fa fa-check"></i><b>2.2</b> Package ecosystem</a>
<ul>
<li class="chapter" data-level="2.2.1" data-path="packages.html"><a href="packages.html#mia-package-family"><i class="fa fa-check"></i><b>2.2.1</b> mia package family</a></li>
<li class="chapter" data-level="2.2.2" data-path="packages.html"><a href="packages.html#sub-diff-abund"><i class="fa fa-check"></i><b>2.2.2</b> Differential abundance</a></li>
<li class="chapter" data-level="2.2.3" data-path="packages.html"><a href="packages.html#other-packages"><i class="fa fa-check"></i><b>2.2.3</b> Other packages</a></li>
<li class="chapter" data-level="2.2.4" data-path="packages.html"><a href="packages.html#open-microbiome-data"><i class="fa fa-check"></i><b>2.2.4</b> Open microbiome data</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="containers.html"><a href="containers.html"><i class="fa fa-check"></i><b>3</b> Microbiome Data</a>
<ul>
<li class="chapter" data-level="3.1" data-path="containers.html"><a href="containers.html#data-science-framework"><i class="fa fa-check"></i><b>3.1</b> Data science framework</a></li>
<li class="chapter" data-level="3.2" data-path="containers.html"><a href="containers.html#data-containers"><i class="fa fa-check"></i><b>3.2</b> Data containers</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="containers.html"><a href="containers.html#assay-slot"><i class="fa fa-check"></i><b>3.2.1</b> Assay data</a></li>
<li class="chapter" data-level="3.2.2" data-path="containers.html"><a href="containers.html#coldata"><i class="fa fa-check"></i><b>3.2.2</b> colData</a></li>
<li class="chapter" data-level="3.2.3" data-path="containers.html"><a href="containers.html#rowdata"><i class="fa fa-check"></i><b>3.2.3</b> rowData</a></li>
<li class="chapter" data-level="3.2.4" data-path="containers.html"><a href="containers.html#rowtree"><i class="fa fa-check"></i><b>3.2.4</b> rowTree</a></li>
<li class="chapter" data-level="3.2.5" data-path="containers.html"><a href="containers.html#alt-exp"><i class="fa fa-check"></i><b>3.2.5</b> Alternative experiments</a></li>
<li class="chapter" data-level="3.2.6" data-path="containers.html"><a href="containers.html#mae"><i class="fa fa-check"></i><b>3.2.6</b> MultiAssayExperiments</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="containers.html"><a href="containers.html#example-data"><i class="fa fa-check"></i><b>3.3</b> Demonstration data</a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="containers.html"><a href="containers.html#package-data"><i class="fa fa-check"></i><b>3.3.1</b> Package data</a></li>
<li class="chapter" data-level="3.3.2" data-path="containers.html"><a href="containers.html#experimenthub-data"><i class="fa fa-check"></i><b>3.3.2</b> ExperimentHub data</a></li>
<li class="chapter" data-level="3.3.3" data-path="containers.html"><a href="containers.html#curated-metagenomic-data"><i class="fa fa-check"></i><b>3.3.3</b> Curated metagenomic data</a></li>
<li class="chapter" data-level="3.3.4" data-path="containers.html"><a href="containers.html#other-data-sources"><i class="fa fa-check"></i><b>3.3.4</b> Other data sources</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="containers.html"><a href="containers.html#loading-experimental-microbiome-data"><i class="fa fa-check"></i><b>3.4</b> Loading experimental microbiome data</a>
<ul>
<li class="chapter" data-level="3.4.1" data-path="containers.html"><a href="containers.html#s-workflow"><i class="fa fa-check"></i><b>3.4.1</b> 16S workflow</a></li>
<li class="chapter" data-level="3.4.2" data-path="containers.html"><a href="containers.html#import-from-file"><i class="fa fa-check"></i><b>3.4.2</b> Import from external files</a></li>
<li class="chapter" data-level="3.4.3" data-path="containers.html"><a href="containers.html#constructing-treesummarizedexperiment"><i class="fa fa-check"></i><b>3.4.3</b> Constructing TreeSummarizedExperiment</a></li>
<li class="chapter" data-level="3.4.4" data-path="containers.html"><a href="containers.html#constructing-multiassayexperiment"><i class="fa fa-check"></i><b>3.4.4</b> Constructing MultiAssayExperiment</a></li>
<li class="chapter" data-level="3.4.5" data-path="containers.html"><a href="containers.html#import-functions-for-standard-formats"><i class="fa fa-check"></i><b>3.4.5</b> Import functions for standard formats</a></li>
<li class="chapter" data-level="3.4.6" data-path="containers.html"><a href="containers.html#conversions-between-data-formats-in-r"><i class="fa fa-check"></i><b>3.4.6</b> Conversions between data formats in R</a></li>
</ul></li>
</ul></li>
<li class="part"><span><b>II Focus Topics</b></span></li>
<li class="chapter" data-level="4" data-path="datamanipulation.html"><a href="datamanipulation.html"><i class="fa fa-check"></i><b>4</b> Data Manipulation</a>
<ul>
<li class="chapter" data-level="4.1" data-path="datamanipulation.html"><a href="datamanipulation.html#tidying-and-subsetting"><i class="fa fa-check"></i><b>4.1</b> Tidying and subsetting</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="datamanipulation.html"><a href="datamanipulation.html#tidy-data"><i class="fa fa-check"></i><b>4.1.1</b> Tidy data</a></li>
<li class="chapter" data-level="4.1.2" data-path="datamanipulation.html"><a href="datamanipulation.html#subsetting"><i class="fa fa-check"></i><b>4.1.2</b> Subsetting</a></li>
<li class="chapter" data-level="4.1.3" data-path="datamanipulation.html"><a href="datamanipulation.html#splitting"><i class="fa fa-check"></i><b>4.1.3</b> Splitting</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="datamanipulation.html"><a href="datamanipulation.html#add-or-modify-data"><i class="fa fa-check"></i><b>4.2</b> Add or modify data</a></li>
<li class="chapter" data-level="4.3" data-path="datamanipulation.html"><a href="datamanipulation.html#merge-data"><i class="fa fa-check"></i><b>4.3</b> Merge data</a>
<ul>
<li class="chapter" data-level="4.3.1" data-path="datamanipulation.html"><a href="datamanipulation.html#additional-functions"><i class="fa fa-check"></i><b>4.3.1</b> Additional functions</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="quality-control.html"><a href="quality-control.html"><i class="fa fa-check"></i><b>5</b> Exploration and Quality Control</a>
<ul>
<li class="chapter" data-level="5.1" data-path="quality-control.html"><a href="quality-control.html#abundance"><i class="fa fa-check"></i><b>5.1</b> Abundance</a></li>
<li class="chapter" data-level="5.2" data-path="quality-control.html"><a href="quality-control.html#prevalence"><i class="fa fa-check"></i><b>5.2</b> Prevalence</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="quality-control.html"><a href="quality-control.html#prevalence-analysis"><i class="fa fa-check"></i><b>5.2.1</b> Prevalence analysis</a></li>
<li class="chapter" data-level="5.2.2" data-path="quality-control.html"><a href="quality-control.html#rare-taxa"><i class="fa fa-check"></i><b>5.2.2</b> Rare taxa</a></li>
<li class="chapter" data-level="5.2.3" data-path="quality-control.html"><a href="quality-control.html#plotting-prevalence"><i class="fa fa-check"></i><b>5.2.3</b> Plotting prevalence</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="quality-control.html"><a href="quality-control.html#qc"><i class="fa fa-check"></i><b>5.3</b> Quality control</a>
<ul>
<li class="chapter" data-level="5.3.1" data-path="quality-control.html"><a href="quality-control.html#top-taxa"><i class="fa fa-check"></i><b>5.3.1</b> Top taxa</a></li>
<li class="chapter" data-level="5.3.2" data-path="quality-control.html"><a href="quality-control.html#library-size-read-count"><i class="fa fa-check"></i><b>5.3.2</b> Library size / read count</a></li>
<li class="chapter" data-level="5.3.3" data-path="quality-control.html"><a href="quality-control.html#contaminant-sequences"><i class="fa fa-check"></i><b>5.3.3</b> Contaminant sequences</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="taxonomic-information.html"><a href="taxonomic-information.html"><i class="fa fa-check"></i><b>6</b> Taxonomic Information</a>
<ul>
<li class="chapter" data-level="6.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#assigning-taxonomic-information."><i class="fa fa-check"></i><b>6.1</b> Assigning taxonomic information.</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#dada2"><i class="fa fa-check"></i><b>6.1.1</b> dada2</a></li>
<li class="chapter" data-level="6.1.2" data-path="taxonomic-information.html"><a href="taxonomic-information.html#decipher"><i class="fa fa-check"></i><b>6.1.2</b> DECIPHER</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="taxonomic-information.html"><a href="taxonomic-information.html#functions-to-access-taxonomic-information"><i class="fa fa-check"></i><b>6.2</b> Functions to access taxonomic information</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#fly-tree"><i class="fa fa-check"></i><b>6.2.1</b> Generate a taxonomic tree on the fly</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="taxonomic-information.html"><a href="taxonomic-information.html#data-agglomeration"><i class="fa fa-check"></i><b>6.3</b> Data agglomeration</a>
<ul>
<li class="chapter" data-level="6.3.1" data-path="taxonomic-information.html"><a href="taxonomic-information.html#taxa-clustering"><i class="fa fa-check"></i><b>6.3.1</b> Taxa clustering</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="taxonomic-information.html"><a href="taxonomic-information.html#assay-transform"><i class="fa fa-check"></i><b>6.4</b> Data transformation</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="community-diversity.html"><a href="community-diversity.html"><i class="fa fa-check"></i><b>7</b> Community Diversity</a>
<ul>
<li class="chapter" data-level="7.1" data-path="community-diversity.html"><a href="community-diversity.html#estimation"><i class="fa fa-check"></i><b>7.1</b> Estimation</a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="community-diversity.html"><a href="community-diversity.html#richness"><i class="fa fa-check"></i><b>7.1.1</b> Richness</a></li>
<li class="chapter" data-level="7.1.2" data-path="community-diversity.html"><a href="community-diversity.html#estimate-diversity"><i class="fa fa-check"></i><b>7.1.2</b> Diversity</a></li>
<li class="chapter" data-level="7.1.3" data-path="community-diversity.html"><a href="community-diversity.html#faith-diversity"><i class="fa fa-check"></i><b>7.1.3</b> Faith phylogenetic diversity</a></li>
<li class="chapter" data-level="7.1.4" data-path="community-diversity.html"><a href="community-diversity.html#evenness"><i class="fa fa-check"></i><b>7.1.4</b> Evenness</a></li>
<li class="chapter" data-level="7.1.5" data-path="community-diversity.html"><a href="community-diversity.html#dominance"><i class="fa fa-check"></i><b>7.1.5</b> Dominance</a></li>
<li class="chapter" data-level="7.1.6" data-path="community-diversity.html"><a href="community-diversity.html#rarity"><i class="fa fa-check"></i><b>7.1.6</b> Rarity</a></li>
<li class="chapter" data-level="7.1.7" data-path="community-diversity.html"><a href="community-diversity.html#divergence"><i class="fa fa-check"></i><b>7.1.7</b> Divergence</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="community-diversity.html"><a href="community-diversity.html#visualization"><i class="fa fa-check"></i><b>7.2</b> Visualization</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="community-similarity.html"><a href="community-similarity.html"><i class="fa fa-check"></i><b>8</b> Community Similarity</a>
<ul>
<li class="chapter" data-level="8.1" data-path="community-similarity.html"><a href="community-similarity.html#unsupervised-ordination"><i class="fa fa-check"></i><b>8.1</b> Unsupervised ordination</a>
<ul>
<li class="chapter" data-level="8.1.1" data-path="community-similarity.html"><a href="community-similarity.html#comparing-communities-by-beta-diversity-analysis"><i class="fa fa-check"></i><b>8.1.1</b> Comparing communities by beta diversity analysis</a></li>
<li class="chapter" data-level="8.1.2" data-path="community-similarity.html"><a href="community-similarity.html#other-ord-methods"><i class="fa fa-check"></i><b>8.1.2</b> Other ordination methods</a></li>
<li class="chapter" data-level="8.1.3" data-path="community-similarity.html"><a href="community-similarity.html#explained-variance"><i class="fa fa-check"></i><b>8.1.3</b> Explained variance</a></li>
</ul></li>
<li class="chapter" data-level="8.2" data-path="community-similarity.html"><a href="community-similarity.html#supervised-ordination"><i class="fa fa-check"></i><b>8.2</b> Supervised ordination</a></li>
<li class="chapter" data-level="8.3" data-path="community-similarity.html"><a href="community-similarity.html#case-studies"><i class="fa fa-check"></i><b>8.3</b> Case studies</a>
<ul>
<li class="chapter" data-level="8.3.1" data-path="community-similarity.html"><a href="community-similarity.html#dbrda-workflow"><i class="fa fa-check"></i><b>8.3.1</b> Testing differences in community composition between sample groups</a></li>
<li class="chapter" data-level="8.3.2" data-path="community-similarity.html"><a href="community-similarity.html#checking-the-homogeneity-condition"><i class="fa fa-check"></i><b>8.3.2</b> Checking the homogeneity condition</a></li>
</ul></li>
<li class="chapter" data-level="8.4" data-path="community-similarity.html"><a href="community-similarity.html#summary"><i class="fa fa-check"></i><b>8.4</b> Summary</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="microbiome-community.html"><a href="microbiome-community.html"><i class="fa fa-check"></i><b>9</b> Community Composition</a>
<ul>
<li class="chapter" data-level="9.1" data-path="microbiome-community.html"><a href="microbiome-community.html#visual-composition"><i class="fa fa-check"></i><b>9.1</b> Visualizing taxonomic composition</a>
<ul>
<li class="chapter" data-level="9.1.1" data-path="microbiome-community.html"><a href="microbiome-community.html#composition-barplot"><i class="fa fa-check"></i><b>9.1.1</b> Composition barplot</a></li>
<li class="chapter" data-level="9.1.2" data-path="microbiome-community.html"><a href="microbiome-community.html#composition-heatmap"><i class="fa fa-check"></i><b>9.1.2</b> Composition heatmap</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="10" data-path="clustering.html"><a href="clustering.html"><i class="fa fa-check"></i><b>10</b> Community Typing (Clustering)</a>
<ul>
<li class="chapter" data-level="10.1" data-path="clustering.html"><a href="clustering.html#custom-tools"><i class="fa fa-check"></i><b>10.1</b> Custom tools</a></li>
<li class="chapter" data-level="10.2" data-path="clustering.html"><a href="clustering.html#hierarchical-clustering"><i class="fa fa-check"></i><b>10.2</b> Hierarchical clustering</a></li>
<li class="chapter" data-level="10.3" data-path="clustering.html"><a href="clustering.html#dirichlet-multinomial-mixtures-dmm"><i class="fa fa-check"></i><b>10.3</b> Dirichlet Multinomial Mixtures (DMM)</a>
<ul>
<li class="chapter" data-level="10.3.1" data-path="clustering.html"><a href="clustering.html#pcoa-with-dmm-clusters"><i class="fa fa-check"></i><b>10.3.1</b> PCoA with DMM clusters</a></li>
</ul></li>
<li class="chapter" data-level="10.4" data-path="clustering.html"><a href="clustering.html#biclustering"><i class="fa fa-check"></i><b>10.4</b> Biclustering</a>
<ul>
<li class="chapter" data-level="10.4.1" data-path="clustering.html"><a href="clustering.html#taxa-vs-samples"><i class="fa fa-check"></i><b>10.4.1</b> Taxa vs samples</a></li>
<li class="chapter" data-level="10.4.2" data-path="clustering.html"><a href="clustering.html#taxa-vs-biomolecules"><i class="fa fa-check"></i><b>10.4.2</b> Taxa vs biomolecules</a></li>
</ul></li>
<li class="chapter" data-level="10.5" data-path="clustering.html"><a href="clustering.html#additional-community-typing"><i class="fa fa-check"></i><b>10.5</b> Additional Community Typing</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="differential-abundance.html"><a href="differential-abundance.html"><i class="fa fa-check"></i><b>11</b> Differential Abundance</a>
<ul>
<li class="chapter" data-level="11.1" data-path="differential-abundance.html"><a href="differential-abundance.html#statistical-challenges-of-microbiome-data"><i class="fa fa-check"></i><b>11.1</b> Statistical challenges of microbiome data</a></li>
<li class="chapter" data-level="11.2" data-path="differential-abundance.html"><a href="differential-abundance.html#using-the-tools"><i class="fa fa-check"></i><b>11.2</b> Using the tools</a>
<ul>
<li class="chapter" data-level="11.2.1" data-path="differential-abundance.html"><a href="differential-abundance.html#preparing-the-data-for-daa"><i class="fa fa-check"></i><b>11.2.1</b> Preparing the data for DAA</a></li>
<li class="chapter" data-level="11.2.2" data-path="differential-abundance.html"><a href="differential-abundance.html#aldex2"><i class="fa fa-check"></i><b>11.2.2</b> ALDEx2</a></li>
<li class="chapter" data-level="11.2.3" data-path="differential-abundance.html"><a href="differential-abundance.html#ancom-bc"><i class="fa fa-check"></i><b>11.2.3</b> ANCOM-BC</a></li>
<li class="chapter" data-level="11.2.4" data-path="differential-abundance.html"><a href="differential-abundance.html#maaslin2"><i class="fa fa-check"></i><b>11.2.4</b> MaAsLin2</a></li>
<li class="chapter" data-level="11.2.5" data-path="differential-abundance.html"><a href="differential-abundance.html#linda"><i class="fa fa-check"></i><b>11.2.5</b> LinDA</a></li>
<li class="chapter" data-level="11.2.6" data-path="differential-abundance.html"><a href="differential-abundance.html#zicoseq"><i class="fa fa-check"></i><b>11.2.6</b> ZicoSeq</a></li>
<li class="chapter" data-level="11.2.7" data-path="differential-abundance.html"><a href="differential-abundance.html#philr"><i class="fa fa-check"></i><b>11.2.7</b> PhILR</a></li>
<li class="chapter" data-level="11.2.8" data-path="differential-abundance.html"><a href="differential-abundance.html#comparison-of-methods"><i class="fa fa-check"></i><b>11.2.8</b> Comparison of methods</a></li>
</ul></li>
<li class="chapter" data-level="11.3" data-path="differential-abundance.html"><a href="differential-abundance.html#daa-with-confounding"><i class="fa fa-check"></i><b>11.3</b> DAA with confounding</a>
<ul>
<li class="chapter" data-level="11.3.1" data-path="differential-abundance.html"><a href="differential-abundance.html#selecting-confounders"><i class="fa fa-check"></i><b>11.3.1</b> Selecting confounders</a></li>
<li class="chapter" data-level="11.3.2" data-path="differential-abundance.html"><a href="differential-abundance.html#ancom-bc-1"><i class="fa fa-check"></i><b>11.3.2</b> ANCOM-BC</a></li>
<li class="chapter" data-level="11.3.3" data-path="differential-abundance.html"><a href="differential-abundance.html#linda-1"><i class="fa fa-check"></i><b>11.3.3</b> LinDA</a></li>
<li class="chapter" data-level="11.3.4" data-path="differential-abundance.html"><a href="differential-abundance.html#zicoseq-1"><i class="fa fa-check"></i><b>11.3.4</b> ZicoSeq</a></li>
</ul></li>
<li class="chapter" data-level="11.4" data-path="differential-abundance.html"><a href="differential-abundance.html#additional-resources"><i class="fa fa-check"></i><b>11.4</b> Additional resources</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="machine_learning.html"><a href="machine_learning.html"><i class="fa fa-check"></i><b>12</b> Machine Learning</a>
<ul>
<li class="chapter" data-level="12.1" data-path="machine_learning.html"><a href="machine_learning.html#supervised-machine-learning"><i class="fa fa-check"></i><b>12.1</b> Supervised machine learning</a></li>
<li class="chapter" data-level="12.2" data-path="machine_learning.html"><a href="machine_learning.html#unsupervised-machine-learning"><i class="fa fa-check"></i><b>12.2</b> Unsupervised machine learning</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="multi-assay-analyses.html"><a href="multi-assay-analyses.html"><i class="fa fa-check"></i><b>13</b> Multi-Assay Analyses</a>
<ul>
<li class="chapter" data-level="13.1" data-path="multi-assay-analyses.html"><a href="multi-assay-analyses.html#cross-correlation"><i class="fa fa-check"></i><b>13.1</b> Cross-correlation Analysis</a></li>
<li class="chapter" data-level="13.2" data-path="multi-assay-analyses.html"><a href="multi-assay-analyses.html#mofa"><i class="fa fa-check"></i><b>13.2</b> Multi-Omics Factor Analysis</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="viz-chapter.html"><a href="viz-chapter.html"><i class="fa fa-check"></i><b>14</b> Visualization</a>
<ul>
<li class="chapter" data-level="14.1" data-path="viz-chapter.html"><a href="viz-chapter.html#pre-analysis-exploration"><i class="fa fa-check"></i><b>14.1</b> Pre-analysis exploration</a>
<ul>
<li class="chapter" data-level="14.1.1" data-path="viz-chapter.html"><a href="viz-chapter.html#accessing-row-and-column-data"><i class="fa fa-check"></i><b>14.1.1</b> Accessing row and column data</a></li>
<li class="chapter" data-level="14.1.2" data-path="viz-chapter.html"><a href="viz-chapter.html#viewing-abundance-and-prevalence-patterns"><i class="fa fa-check"></i><b>14.1.2</b> Viewing abundance and prevalence patterns</a></li>
</ul></li>
<li class="chapter" data-level="14.2" data-path="viz-chapter.html"><a href="viz-chapter.html#diversity-estimation"><i class="fa fa-check"></i><b>14.2</b> Diversity estimation</a>
<ul>
<li class="chapter" data-level="14.2.1" data-path="viz-chapter.html"><a href="viz-chapter.html#alpha-diversity-with-scatter-violin-and-box-plots"><i class="fa fa-check"></i><b>14.2.1</b> Alpha diversity with scatter, violin and box plots</a></li>
<li class="chapter" data-level="14.2.2" data-path="viz-chapter.html"><a href="viz-chapter.html#beta-diversity-with-shepard-and-coordination-plots"><i class="fa fa-check"></i><b>14.2.2</b> Beta diversity with Shepard and coordination plots</a></li>
</ul></li>
<li class="chapter" data-level="14.3" data-path="viz-chapter.html"><a href="viz-chapter.html#statistical-analysis"><i class="fa fa-check"></i><b>14.3</b> Statistical analysis</a>
<ul>
<li class="chapter" data-level="14.3.1" data-path="viz-chapter.html"><a href="viz-chapter.html#heatmaps"><i class="fa fa-check"></i><b>14.3.1</b> Heatmaps</a></li>
</ul></li>
</ul></li>
<li class="part"><span><b>III Training</b></span></li>
<li class="chapter" data-level="15" data-path="training.html"><a href="training.html"><i class="fa fa-check"></i><b>15</b> Training</a>
<ul>
<li class="chapter" data-level="15.1" data-path="training.html"><a href="training.html#checklist"><i class="fa fa-check"></i><b>15.1</b> Checklist</a></li>
<li class="chapter" data-level="15.2" data-path="training.html"><a href="training.html#software"><i class="fa fa-check"></i><b>15.2</b> Recommended software</a></li>
<li class="chapter" data-level="15.3" data-path="training.html"><a href="training.html#material"><i class="fa fa-check"></i><b>15.3</b> Study material</a></li>
<li class="chapter" data-level="15.4" data-path="training.html"><a href="training.html#support-and-resources"><i class="fa fa-check"></i><b>15.4</b> Support and resources</a></li>
<li class="chapter" data-level="15.5" data-path="training.html"><a href="training.html#further-reading"><i class="fa fa-check"></i><b>15.5</b> Further reading</a></li>
<li class="chapter" data-level="15.6" data-path="training.html"><a href="training.html#coc"><i class="fa fa-check"></i><b>15.6</b> Code of Conduct</a></li>
</ul></li>
<li class="chapter" data-level="16" data-path="resources.html"><a href="resources.html"><i class="fa fa-check"></i><b>16</b> Resources</a>
<ul>
<li class="chapter" data-level="16.1" data-path="resources.html"><a href="resources.html#data-containers-1"><i class="fa fa-check"></i><b>16.1</b> Data containers</a>
<ul>
<li class="chapter" data-level="16.1.1" data-path="resources.html"><a href="resources.html#data-container-documentation"><i class="fa fa-check"></i><b>16.1.1</b> Data container documentation</a></li>
<li class="chapter" data-level="16.1.2" data-path="resources.html"><a href="resources.html#other-relevant-containers"><i class="fa fa-check"></i><b>16.1.2</b> Other relevant containers</a></li>
<li class="chapter" data-level="16.1.3" data-path="resources.html"><a href="resources.html#phyloseq-an-alternative-container-for-microbiome-data"><i class="fa fa-check"></i><b>16.1.3</b> phyloseq: an alternative container for microbiome data</a></li>
</ul></li>
<li class="chapter" data-level="16.2" data-path="resources.html"><a href="resources.html#r-programming-resources"><i class="fa fa-check"></i><b>16.2</b> R programming resources</a>
<ul>
<li class="chapter" data-level="16.2.1" data-path="resources.html"><a href="resources.html#base-r-and-rstudio"><i class="fa fa-check"></i><b>16.2.1</b> Base R and RStudio</a></li>
<li class="chapter" data-level="16.2.2" data-path="resources.html"><a href="resources.html#bioc_intro"><i class="fa fa-check"></i><b>16.2.2</b> Bioconductor Classes</a></li>
</ul></li>
<li class="chapter" data-level="16.3" data-path="resources.html"><a href="resources.html#quarto"><i class="fa fa-check"></i><b>16.3</b> Reproducible reporting with Quarto</a>
<ul>
<li class="chapter" data-level="16.3.1" data-path="resources.html"><a href="resources.html#learn-quarto"><i class="fa fa-check"></i><b>16.3.1</b> Learn Quarto</a></li>
<li class="chapter" data-level="16.3.2" data-path="resources.html"><a href="resources.html#additional-material-on-rmarkdown"><i class="fa fa-check"></i><b>16.3.2</b> Additional material on Rmarkdown</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="17" data-path="exercises.html"><a href="exercises.html"><i class="fa fa-check"></i><b>17</b> Exercises</a>
<ul>
<li class="chapter" data-level="17.1" data-path="exercises.html"><a href="exercises.html#basics-of-rbioconductor"><i class="fa fa-check"></i><b>17.1</b> Basics of R/Bioconductor</a></li>
<li class="chapter" data-level="17.2" data-path="exercises.html"><a href="exercises.html#workflows"><i class="fa fa-check"></i><b>17.2</b> Workflows</a>
<ul>
<li class="chapter" data-level="17.2.1" data-path="exercises.html"><a href="exercises.html#reproducible-reporting-with-quarto"><i class="fa fa-check"></i><b>17.2.1</b> Reproducible reporting with Quarto</a></li>
</ul></li>
<li class="chapter" data-level="17.3" data-path="exercises.html"><a href="exercises.html#data-containers-treese"><i class="fa fa-check"></i><b>17.3</b> Data containers: TreeSE</a>
<ul>
<li class="chapter" data-level="17.3.1" data-path="exercises.html"><a href="exercises.html#construct-TreeSE"><i class="fa fa-check"></i><b>17.3.1</b> Constructing a data object</a></li>
<li class="chapter" data-level="17.3.2" data-path="exercises.html"><a href="exercises.html#importing-data"><i class="fa fa-check"></i><b>17.3.2</b> Importing data</a></li>
<li class="chapter" data-level="17.3.3" data-path="exercises.html"><a href="exercises.html#preliminary-exploration"><i class="fa fa-check"></i><b>17.3.3</b> Preliminary exploration</a></li>
<li class="chapter" data-level="17.3.4" data-path="exercises.html"><a href="exercises.html#assay-retrieval"><i class="fa fa-check"></i><b>17.3.4</b> Assay retrieval</a></li>
<li class="chapter" data-level="17.3.5" data-path="exercises.html"><a href="exercises.html#sample-information"><i class="fa fa-check"></i><b>17.3.5</b> Sample information</a></li>
<li class="chapter" data-level="17.3.6" data-path="exercises.html"><a href="exercises.html#feature-information"><i class="fa fa-check"></i><b>17.3.6</b> Feature information</a></li>
<li class="chapter" data-level="17.3.7" data-path="exercises.html"><a href="exercises.html#other-elements"><i class="fa fa-check"></i><b>17.3.7</b> Other elements</a></li>
</ul></li>
<li class="chapter" data-level="17.4" data-path="exercises.html"><a href="exercises.html#data-manipulation"><i class="fa fa-check"></i><b>17.4</b> Data manipulation</a>
<ul>
<li class="chapter" data-level="17.4.1" data-path="exercises.html"><a href="exercises.html#subsetting-1"><i class="fa fa-check"></i><b>17.4.1</b> Subsetting</a></li>
<li class="chapter" data-level="17.4.2" data-path="exercises.html"><a href="exercises.html#library-sizes"><i class="fa fa-check"></i><b>17.4.2</b> Library sizes</a></li>
<li class="chapter" data-level="17.4.3" data-path="exercises.html"><a href="exercises.html#prevalent-and-core-taxonomic-features"><i class="fa fa-check"></i><b>17.4.3</b> Prevalent and core taxonomic features</a></li>
<li class="chapter" data-level="17.4.4" data-path="exercises.html"><a href="exercises.html#data-exploration"><i class="fa fa-check"></i><b>17.4.4</b> Data exploration</a></li>
<li class="chapter" data-level="17.4.5" data-path="exercises.html"><a href="exercises.html#other-functions"><i class="fa fa-check"></i><b>17.4.5</b> Other functions</a></li>
<li class="chapter" data-level="17.4.6" data-path="exercises.html"><a href="exercises.html#assay-transformation"><i class="fa fa-check"></i><b>17.4.6</b> Assay transformation</a></li>
</ul></li>
<li class="chapter" data-level="17.5" data-path="exercises.html"><a href="exercises.html#abundance-tables"><i class="fa fa-check"></i><b>17.5</b> Abundance tables</a>
<ul>
<li class="chapter" data-level="17.5.1" data-path="exercises.html"><a href="exercises.html#taxonomic-levels"><i class="fa fa-check"></i><b>17.5.1</b> Taxonomic levels</a></li>
<li class="chapter" data-level="17.5.2" data-path="exercises.html"><a href="exercises.html#alternative-experiments"><i class="fa fa-check"></i><b>17.5.2</b> Alternative experiments</a></li>
</ul></li>
<li class="chapter" data-level="17.6" data-path="exercises.html"><a href="exercises.html#community-alpha-diversity"><i class="fa fa-check"></i><b>17.6</b> Community (alpha) diversity</a>
<ul>
<li class="chapter" data-level="17.6.1" data-path="exercises.html"><a href="exercises.html#estimation-1"><i class="fa fa-check"></i><b>17.6.1</b> Estimation</a></li>
<li class="chapter" data-level="17.6.2" data-path="exercises.html"><a href="exercises.html#visualization-1"><i class="fa fa-check"></i><b>17.6.2</b> Visualization</a></li>
<li class="chapter" data-level="17.6.3" data-path="exercises.html"><a href="exercises.html#correlation"><i class="fa fa-check"></i><b>17.6.3</b> Correlation</a></li>
<li class="chapter" data-level="17.6.4" data-path="exercises.html"><a href="exercises.html#differences-between-groups"><i class="fa fa-check"></i><b>17.6.4</b> Differences between groups</a></li>
</ul></li>
<li class="chapter" data-level="17.7" data-path="exercises.html"><a href="exercises.html#community-similarity-1"><i class="fa fa-check"></i><b>17.7</b> Community similarity</a>
<ul>
<li class="chapter" data-level="17.7.1" data-path="exercises.html"><a href="exercises.html#reduced-dimensions-retrieval"><i class="fa fa-check"></i><b>17.7.1</b> Reduced dimensions retrieval</a></li>
<li class="chapter" data-level="17.7.2" data-path="exercises.html"><a href="exercises.html#visualization-basics-with-pca"><i class="fa fa-check"></i><b>17.7.2</b> Visualization basics with PCA</a></li>
<li class="chapter" data-level="17.7.3" data-path="exercises.html"><a href="exercises.html#principal-coordinate-analysis-pcoa"><i class="fa fa-check"></i><b>17.7.3</b> Principal Coordinate Analysis (PCoA)</a></li>
<li class="chapter" data-level="17.7.4" data-path="exercises.html"><a href="exercises.html#permanova-analysis"><i class="fa fa-check"></i><b>17.7.4</b> PERMANOVA analysis</a></li>
<li class="chapter" data-level="17.7.5" data-path="exercises.html"><a href="exercises.html#redundancy-analysis-rda"><i class="fa fa-check"></i><b>17.7.5</b> Redundancy analysis (RDA)</a></li>
<li class="chapter" data-level="17.7.6" data-path="exercises.html"><a href="exercises.html#beta-diversity-analysis"><i class="fa fa-check"></i><b>17.7.6</b> Beta diversity analysis</a></li>
</ul></li>
<li class="chapter" data-level="17.8" data-path="exercises.html"><a href="exercises.html#differential-abundance-1"><i class="fa fa-check"></i><b>17.8</b> Differential abundance</a>
<ul>
<li class="chapter" data-level="17.8.1" data-path="exercises.html"><a href="exercises.html#standard-analysis-with-aldex2"><i class="fa fa-check"></i><b>17.8.1</b> Standard analysis with ALDEx2</a></li>
<li class="chapter" data-level="17.8.2" data-path="exercises.html"><a href="exercises.html#control-daa-confounders"><i class="fa fa-check"></i><b>17.8.2</b> Controlling for confounders</a></li>
<li class="chapter" data-level="17.8.3" data-path="exercises.html"><a href="exercises.html#compare-daa-methods"><i class="fa fa-check"></i><b>17.8.3</b> Comparing methods</a></li>
<li class="chapter" data-level="17.8.4" data-path="exercises.html"><a href="exercises.html#workflow-1"><i class="fa fa-check"></i><b>17.8.4</b> Workflow 1</a></li>
<li class="chapter" data-level="17.8.5" data-path="exercises.html"><a href="exercises.html#workflow-2"><i class="fa fa-check"></i><b>17.8.5</b> Workflow 2</a></li>
</ul></li>
<li class="chapter" data-level="17.9" data-path="exercises.html"><a href="exercises.html#visualization-2"><i class="fa fa-check"></i><b>17.9</b> Visualization</a>
<ul>
<li class="chapter" data-level="17.9.1" data-path="exercises.html"><a href="exercises.html#multivariate-ordination"><i class="fa fa-check"></i><b>17.9.1</b> Multivariate ordination</a></li>
<li class="chapter" data-level="17.9.2" data-path="exercises.html"><a href="exercises.html#heatmap-visualization"><i class="fa fa-check"></i><b>17.9.2</b> Heatmap visualization</a></li>
</ul></li>
<li class="chapter" data-level="17.10" data-path="exercises.html"><a href="exercises.html#multiomics"><i class="fa fa-check"></i><b>17.10</b> Multiomics</a>
<ul>
<li class="chapter" data-level="17.10.1" data-path="exercises.html"><a href="exercises.html#basic-exploration"><i class="fa fa-check"></i><b>17.10.1</b> Basic exploration</a></li>
<li class="chapter" data-level="17.10.2" data-path="exercises.html"><a href="exercises.html#experiment-agglomeration"><i class="fa fa-check"></i><b>17.10.2</b> Experiment agglomeration</a></li>
<li class="chapter" data-level="17.10.3" data-path="exercises.html"><a href="exercises.html#experiment-transformation"><i class="fa fa-check"></i><b>17.10.3</b> Experiment transformation</a></li>
<li class="chapter" data-level="17.10.4" data-path="exercises.html"><a href="exercises.html#assay-extraction"><i class="fa fa-check"></i><b>17.10.4</b> Assay extraction</a></li>
<li class="chapter" data-level="17.10.5" data-path="exercises.html"><a href="exercises.html#mae-reconstruction"><i class="fa fa-check"></i><b>17.10.5</b> MAE reconstruction</a></li>
<li class="chapter" data-level="17.10.6" data-path="exercises.html"><a href="exercises.html#cross-correlation-analysis"><i class="fa fa-check"></i><b>17.10.6</b> Cross-correlation analysis</a></li>
</ul></li>
</ul></li>
<li class="part"><span><b>IV Appendix</b></span></li>
<li class="chapter" data-level="18" data-path="extras.html"><a href="extras.html"><i class="fa fa-check"></i><b>18</b> Extra material</a>
<ul>
<li class="chapter" data-level="18.1" data-path="extras.html"><a href="extras.html#slides"><i class="fa fa-check"></i><b>18.1</b> Slides</a></li>
<li class="chapter" data-level="18.2" data-path="extras.html"><a href="extras.html#compare-permanova"><i class="fa fa-check"></i><b>18.2</b> PERMANOVA comparison</a></li>
<li class="chapter" data-level="18.3" data-path="extras.html"><a href="extras.html#bayesian-multinomial-logistic-normal-models"><i class="fa fa-check"></i><b>18.3</b> Bayesian Multinomial Logistic-Normal Models</a></li>
<li class="chapter" data-level="18.4" data-path="extras.html"><a href="extras.html#interactive-3d-plots"><i class="fa fa-check"></i><b>18.4</b> Interactive 3D Plots</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="developers.html"><a href="developers.html"><i class="fa fa-check"></i>Developers</a>
<ul>
<li class="chapter" data-level="" data-path="developers.html"><a href="developers.html#core-team"><i class="fa fa-check"></i>Core team</a></li>
<li class="chapter" data-level="" data-path="developers.html"><a href="developers.html#contributors"><i class="fa fa-check"></i>Contributors</a></li>
<li class="chapter" data-level="" data-path="developers.html"><a href="developers.html#acknowledgments"><i class="fa fa-check"></i>Acknowledgments</a></li>
<li class="chapter" data-level="18.4.1" data-path="developers.html"><a href="developers.html#how-to-contribute"><i class="fa fa-check"></i><b>18.4.1</b> How to contribute</a></li>
<li class="chapter" data-level="" data-path="developers.html"><a href="developers.html#support"><i class="fa fa-check"></i>Support</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="sessioninfo.html"><a href="sessioninfo.html"><i class="fa fa-check"></i>Sessioninfo</a></li>
<li class="chapter" data-level="" data-path="bibliography.html"><a href="bibliography.html"><i class="fa fa-check"></i>Bibliography</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Orchestrating Microbiome Analysis with Bioconductor</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="viz-chapter" class="section level1 hasAnchor" number="14">
<h1><span class="header-section-number">Chapter 14</span> Visualization<a href="viz-chapter.html#viz-chapter" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<script>
document.addEventListener("click", function (event) {
    if (event.target.classList.contains("rebook-collapse")) {
        event.target.classList.toggle("active");
        var content = event.target.nextElementSibling;
        if (content.style.display === "block") {
            content.style.display = "none";
        } else {
            content.style.display = "block";
        }
    }
})
</script>
<style>
.rebook-collapse {
  background-color: #eee;
  color: #444;
  cursor: pointer;
  padding: 18px;
  width: 100%;
  border: none;
  text-align: left;
  outline: none;
  font-size: 15px;
}

.rebook-content {
  padding: 0 18px;
  display: none;
  overflow: hidden;
  background-color: #f1f1f1;
}
</style>
<p>Data visualization will inevitably shape interpretation and motivate
the next steps of the analysis. A variety of visualization methods are
available for microbiome analysis but the application requires careful
attention to details. Knowledge on the available tools and their
limitations plays an important role in selecting the most suitable
methods to address a given question.</p>
<p>This chapter introduces the reader to a number of visualization
techniques found in this book, such as:</p>
<ul>
<li>barplots</li>
<li>boxplots</li>
<li>heatmaps</li>
<li>ordination charts</li>
<li>regression charts</li>
<li>trees</li>
</ul>
<p>The toolkit which provides the essential plotting functionality
includes the following packages:</p>
<ul>
<li><em>patchwork</em>, <em>cowplot</em>, <em>ggpubr</em> and <em>gridExtra</em>: plot layout and multi-panel plotting</li>
<li><em>miaViz</em>: specific visualization tools for <code>TreeSummaizedExperiment</code> objects</li>
<li><em>scater</em>: specific visualization tools for <code>SingleCellExperiment</code> objects</li>
<li><em>ggplot2</em>, <em>pheatmap</em>, <em>ggtree</em>, <em>sechm</em>: composition heatmaps</li>
<li><em>ANCOMBC</em>, <em>ALDEx2</em> and <em>Maaslin2</em>: visual differential abundance</li>
<li><em>fido</em>: tree-based methods for differential abundance</li>
<li><em>plotly</em>: animated and 3D plotting</li>
</ul>
<p>For systematic and extensive tutorials on the visual tools available
in <em>mia</em>, readers can refer to the following material:</p>
<ul>
<li><a href="https://microbiome.github.io/tutorials/">microbiome tutorials</a></li>
</ul>
<div id="pre-analysis-exploration" class="section level2 hasAnchor" number="14.1">
<h2><span class="header-section-number">14.1</span> Pre-analysis exploration<a href="viz-chapter.html#pre-analysis-exploration" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="accessing-row-and-column-data" class="section level3 hasAnchor" number="14.1.1">
<h3><span class="header-section-number">14.1.1</span> Accessing row and column data<a href="viz-chapter.html#accessing-row-and-column-data" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><code>SCE</code> and <code>TreeSE</code> objects contain multiple layers of information in the
form of rows, columns and meta data. The <em>scater</em> package supports in
accessing, modifying and graphing the meta data related to features as
well as samples.</p>
<div class="sourceCode" id="cb385"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb385-1"><a href="viz-chapter.html#cb385-1" tabindex="-1"></a><span class="co"># list row meta data</span></span>
<span id="cb385-2"><a href="viz-chapter.html#cb385-2" tabindex="-1"></a><span class="fu">names</span>(<span class="fu">rowData</span>(tse))</span></code></pre></div>
<pre><code>## [1] &quot;Kingdom&quot; &quot;Phylum&quot;  &quot;Class&quot;   &quot;Order&quot;   &quot;Family&quot;  &quot;Genus&quot;   &quot;Species&quot;</code></pre>
<div class="sourceCode" id="cb387"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb387-1"><a href="viz-chapter.html#cb387-1" tabindex="-1"></a><span class="co"># list column meta data</span></span>
<span id="cb387-2"><a href="viz-chapter.html#cb387-2" tabindex="-1"></a><span class="fu">names</span>(<span class="fu">colData</span>(tse))</span></code></pre></div>
<pre><code>## [1] &quot;X.SampleID&quot;               &quot;Primer&quot;                  
## [3] &quot;Final_Barcode&quot;            &quot;Barcode_truncated_plus_T&quot;
## [5] &quot;Barcode_full_length&quot;      &quot;SampleType&quot;              
## [7] &quot;Description&quot;</code></pre>
<p>Such meta data can be directly plotted with the functions
<code>plotRowData</code> and <code>plotColData</code>.</p>
<div class="sourceCode" id="cb389"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb389-1"><a href="viz-chapter.html#cb389-1" tabindex="-1"></a><span class="co"># obtain QC data</span></span>
<span id="cb389-2"><a href="viz-chapter.html#cb389-2" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">addPerCellQC</span>(tse)</span>
<span id="cb389-3"><a href="viz-chapter.html#cb389-3" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">addPerFeatureQC</span>(tse)</span>
<span id="cb389-4"><a href="viz-chapter.html#cb389-4" tabindex="-1"></a><span class="co"># plot QC Mean against Species</span></span>
<span id="cb389-5"><a href="viz-chapter.html#cb389-5" tabindex="-1"></a><span class="fu">plotRowData</span>(tse, <span class="st">&quot;mean&quot;</span>, <span class="st">&quot;Species&quot;</span>) <span class="sc">+</span></span>
<span id="cb389-6"><a href="viz-chapter.html#cb389-6" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_blank</span>()) <span class="sc">+</span></span>
<span id="cb389-7"><a href="viz-chapter.html#cb389-7" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">&quot;Species&quot;</span>, <span class="at">y =</span> <span class="st">&quot;QC Mean&quot;</span>)</span></code></pre></div>
<p><img src="19_visualization_techniques_files/figure-html/unnamed-chunk-3-1.png" width="672" /></p>
<div class="sourceCode" id="cb390"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb390-1"><a href="viz-chapter.html#cb390-1" tabindex="-1"></a><span class="co"># plot QC Sum against Sample ID, colour-labeled by Sample Type</span></span>
<span id="cb390-2"><a href="viz-chapter.html#cb390-2" tabindex="-1"></a><span class="fu">plotColData</span>(tse, <span class="st">&quot;sum&quot;</span>, <span class="st">&quot;X.SampleID&quot;</span>, <span class="at">colour_by =</span> <span class="st">&quot;SampleType&quot;</span>) <span class="sc">+</span></span>
<span id="cb390-3"><a href="viz-chapter.html#cb390-3" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">45</span>, <span class="at">hjust =</span> <span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb390-4"><a href="viz-chapter.html#cb390-4" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">&quot;Sample ID&quot;</span>, <span class="at">y =</span> <span class="st">&quot;QC Sum&quot;</span>)</span></code></pre></div>
<p><img src="19_visualization_techniques_files/figure-html/unnamed-chunk-3-2.png" width="672" /></p>
<p>Alternatively, they can be converted to a <code>data.frame</code> object and
passed to <code>ggplot</code>.</p>
<div class="sourceCode" id="cb391"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb391-1"><a href="viz-chapter.html#cb391-1" tabindex="-1"></a><span class="co"># store colData into a data frame</span></span>
<span id="cb391-2"><a href="viz-chapter.html#cb391-2" tabindex="-1"></a>coldata <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">colData</span>(tse))</span>
<span id="cb391-3"><a href="viz-chapter.html#cb391-3" tabindex="-1"></a><span class="co"># plot Number of Samples against Sampling Site</span></span>
<span id="cb391-4"><a href="viz-chapter.html#cb391-4" tabindex="-1"></a><span class="fu">ggplot</span>(coldata, <span class="fu">aes</span>(<span class="at">x =</span> SampleType)) <span class="sc">+</span></span>
<span id="cb391-5"><a href="viz-chapter.html#cb391-5" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">width =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb391-6"><a href="viz-chapter.html#cb391-6" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">45</span>, <span class="at">hjust =</span> <span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb391-7"><a href="viz-chapter.html#cb391-7" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">&quot;Sampling Site&quot;</span>,</span>
<span id="cb391-8"><a href="viz-chapter.html#cb391-8" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">&quot;Number of Samples&quot;</span>)</span></code></pre></div>
<p><img src="19_visualization_techniques_files/figure-html/unnamed-chunk-4-1.png" width="672" /></p>
<p>Further methods of application can be found in the chapters <a href="quality-control.html#qc">5.3</a>
and <a href="community-diversity.html#richness">7.1.1</a> and in a few <a href="https://github.com/davismcc/scater_tutorials_open_data">external
tutorials</a>
with open data. Additionally, <code>rowData</code> and <code>colData</code> allow
manipulation and subsetting of large data sets into smaller units, as
explained in chapter <a href="datamanipulation.html#datamanipulation">4</a>.</p>
</div>
<div id="viewing-abundance-and-prevalence-patterns" class="section level3 hasAnchor" number="14.1.2">
<h3><span class="header-section-number">14.1.2</span> Viewing abundance and prevalence patterns<a href="viz-chapter.html#viewing-abundance-and-prevalence-patterns" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Prior-to-analysis exploration may involve questions such as how microorganisms
are distributed across samples (abundance) and what microorganisms are present
in most of the samples (prevalence). The information on abundance and prevalence
can be summarized into a <strong>jitter</strong> or <strong>density plot</strong> and a <strong>tree</strong>,
respectively, with the <em>miaViz</em> package.</p>
<p>Specifically, the functions <code>plotAbundance</code>, <code>plotAbundanceDensity</code>
and <code>plotRowTree</code> are used, and examples on their usage are discussed
throughout chapter <a href="quality-control.html#quality-control">5</a>.</p>
</div>
</div>
<div id="diversity-estimation" class="section level2 hasAnchor" number="14.2">
<h2><span class="header-section-number">14.2</span> Diversity estimation<a href="viz-chapter.html#diversity-estimation" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Alpha diversity is commonly measured as one of the diversity indices
explained in chapter <a href="community-diversity.html#community-diversity">7</a>. Because the focus
lies on each sample separately, one-dimensional plots, such as
<strong>scatter</strong>, <strong>violin</strong> and <strong>box plots</strong>, are suitable.</p>
<p>Beta diversity is generally evaluated as one of the dissimilarity
indices reported in chapter <a href="community-similarity.html#community-similarity">8</a>. Unlike alpha
diversity, samples are compared collectively to estimate the
heterogeneity across them, therefore multidimensional plots, such as
<strong>Shepard</strong> and <strong>ordination plots</strong> are suitable.</p>
<table>
<colgroup>
<col width="31%" />
<col width="35%" />
<col width="33%" />
</colgroup>
<thead>
<tr class="header">
<th align="center"></th>
<th align="center">alpha diversity</th>
<th align="center">beta diversity</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center">used metrics</td>
<td align="center">diversity indices</td>
<td align="center">dissimilarity indices</td>
</tr>
<tr class="even">
<td align="center">metric dimensionality</td>
<td align="center">one-dimensional</td>
<td align="center">multidimensional</td>
</tr>
<tr class="odd">
<td align="center">suitable visualization</td>
<td align="center">scatter, violin, box plots</td>
<td align="center">Shepard, ordination plots</td>
</tr>
</tbody>
</table>
<p>In conclusion, visualization techniques for alpha and beta diversity
significantly differ from one another.</p>
<div id="alpha-diversity-with-scatter-violin-and-box-plots" class="section level3 hasAnchor" number="14.2.1">
<h3><span class="header-section-number">14.2.1</span> Alpha diversity with scatter, violin and box plots<a href="viz-chapter.html#alpha-diversity-with-scatter-violin-and-box-plots" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The basic method to visualize the diversity values assigned to the
different samples in a <code>TSE</code> object includes the following, where each
data point represents one sample:</p>
<div class="sourceCode" id="cb392"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb392-1"><a href="viz-chapter.html#cb392-1" tabindex="-1"></a><span class="co"># estimate shannon diversity index</span></span>
<span id="cb392-2"><a href="viz-chapter.html#cb392-2" tabindex="-1"></a>tse <span class="ot">&lt;-</span> mia<span class="sc">::</span><span class="fu">estimateDiversity</span>(tse, </span>
<span id="cb392-3"><a href="viz-chapter.html#cb392-3" tabindex="-1"></a>                              <span class="at">assay.type =</span> <span class="st">&quot;counts&quot;</span>,</span>
<span id="cb392-4"><a href="viz-chapter.html#cb392-4" tabindex="-1"></a>                              <span class="at">index =</span> <span class="st">&quot;shannon&quot;</span>, </span>
<span id="cb392-5"><a href="viz-chapter.html#cb392-5" tabindex="-1"></a>                              <span class="at">name =</span> <span class="st">&quot;shannon&quot;</span>)</span>
<span id="cb392-6"><a href="viz-chapter.html#cb392-6" tabindex="-1"></a><span class="co"># plot shannon diversity index, colour-labeled by Sample Type</span></span>
<span id="cb392-7"><a href="viz-chapter.html#cb392-7" tabindex="-1"></a><span class="fu">plotColData</span>(tse, <span class="st">&quot;shannon&quot;</span>, <span class="at">colour_by =</span> <span class="st">&quot;SampleType&quot;</span>)</span></code></pre></div>
<p><img src="19_visualization_techniques_files/figure-html/unnamed-chunk-5-1.png" width="672" /></p>
<p>The several indices available for the evaluation of alpha diversity
often return slightly divergent results, which can be visually
compared with a multiple violin or box plot. For this purpose,
<code>plotColData</code> (for violin plots) or <code>ggplot</code> (for box plots) are
recursively applied to a number of diversity indices with the function
<code>lapply</code> and the multi-panel plotting functionality of the <em>patchwork</em>
package is then exploited.</p>
<div class="sourceCode" id="cb393"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb393-1"><a href="viz-chapter.html#cb393-1" tabindex="-1"></a><span class="co"># estimate faith diversity index</span></span>
<span id="cb393-2"><a href="viz-chapter.html#cb393-2" tabindex="-1"></a>tse <span class="ot">&lt;-</span> mia<span class="sc">::</span><span class="fu">estimateFaith</span>(tse,</span>
<span id="cb393-3"><a href="viz-chapter.html#cb393-3" tabindex="-1"></a>                          <span class="at">assay.type =</span> <span class="st">&quot;counts&quot;</span>)</span>
<span id="cb393-4"><a href="viz-chapter.html#cb393-4" tabindex="-1"></a><span class="co"># store colData into a data frame</span></span>
<span id="cb393-5"><a href="viz-chapter.html#cb393-5" tabindex="-1"></a>coldata <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">colData</span>(tse))</span>
<span id="cb393-6"><a href="viz-chapter.html#cb393-6" tabindex="-1"></a><span class="co"># generate plots for shannon and faith indices</span></span>
<span id="cb393-7"><a href="viz-chapter.html#cb393-7" tabindex="-1"></a><span class="co"># and store them into a list</span></span>
<span id="cb393-8"><a href="viz-chapter.html#cb393-8" tabindex="-1"></a>plots <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">c</span>(<span class="st">&quot;shannon&quot;</span>, <span class="st">&quot;faith&quot;</span>),</span>
<span id="cb393-9"><a href="viz-chapter.html#cb393-9" tabindex="-1"></a>                <span class="cf">function</span>(i) <span class="fu">ggplot</span>(coldata, <span class="fu">aes_string</span>(<span class="at">y =</span> i)) <span class="sc">+</span></span>
<span id="cb393-10"><a href="viz-chapter.html#cb393-10" tabindex="-1"></a>                  <span class="fu">geom_boxplot</span>() <span class="sc">+</span></span>
<span id="cb393-11"><a href="viz-chapter.html#cb393-11" tabindex="-1"></a>                  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb393-12"><a href="viz-chapter.html#cb393-12" tabindex="-1"></a>                        <span class="at">axis.ticks.x =</span> <span class="fu">element_blank</span>()))</span>
<span id="cb393-13"><a href="viz-chapter.html#cb393-13" tabindex="-1"></a><span class="co"># combine plots with patchwork</span></span>
<span id="cb393-14"><a href="viz-chapter.html#cb393-14" tabindex="-1"></a>plots[[<span class="dv">1</span>]] <span class="sc">+</span> plots[[<span class="dv">2</span>]]</span></code></pre></div>
<p><img src="19_visualization_techniques_files/figure-html/unnamed-chunk-6-1.png" width="672" /></p>
<p>The analogous output in the form of a violin plot is obtained in
chapter <a href="community-diversity.html#faith-diversity">7.1.3</a>. In addition, box plots that group
samples according to certain information, such as origin, sex, age and
health condition, can be labeled with p-values for significant
differences with the package <em>ggsignif</em> package, as shown in chapter
<a href="community-diversity.html#estimate-diversity">7.1.2</a>.</p>
</div>
<div id="beta-diversity-with-shepard-and-coordination-plots" class="section level3 hasAnchor" number="14.2.2">
<h3><span class="header-section-number">14.2.2</span> Beta diversity with Shepard and coordination plots<a href="viz-chapter.html#beta-diversity-with-shepard-and-coordination-plots" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The <em>scater</em> package offers the general function <code>plotReducedDim</code>. In
its basic form, it takes a <code>TSE</code> object and the results on sample
similarity stored in the same object, which can be evaluated with the
following coordination methods:</p>
<ul>
<li><code>runMDS</code></li>
<li><code>runNMDS</code></li>
<li><code>runPCA</code></li>
<li><code>runTSNE</code></li>
<li><code>runUMAP</code></li>
</ul>
<p>Since these clustering techniques allow for multiple coordinates or
components, <strong>coordination plots</strong> can also span multiple dimensions,
which is explained in chapter <a href="extras.html#extras">18</a>.</p>
<div class="sourceCode" id="cb394"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb394-1"><a href="viz-chapter.html#cb394-1" tabindex="-1"></a><span class="co"># perform NMDS coordination method</span></span>
<span id="cb394-2"><a href="viz-chapter.html#cb394-2" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">runNMDS</span>(tse,</span>
<span id="cb394-3"><a href="viz-chapter.html#cb394-3" tabindex="-1"></a>               <span class="at">FUN =</span> vegan<span class="sc">::</span>vegdist,</span>
<span id="cb394-4"><a href="viz-chapter.html#cb394-4" tabindex="-1"></a>               <span class="at">name =</span> <span class="st">&quot;NMDS&quot;</span>)</span></code></pre></div>
<pre><code>## initial  value 47.733208 
## iter   5 value 33.853364
## iter  10 value 32.891200
## final  value 32.823570 
## converged</code></pre>
<div class="sourceCode" id="cb396"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb396-1"><a href="viz-chapter.html#cb396-1" tabindex="-1"></a><span class="co"># plot results of a 2-component NMDS on tse,</span></span>
<span id="cb396-2"><a href="viz-chapter.html#cb396-2" tabindex="-1"></a><span class="co"># coloured-scaled by shannon diversity index</span></span>
<span id="cb396-3"><a href="viz-chapter.html#cb396-3" tabindex="-1"></a><span class="fu">plotReducedDim</span>(tse, <span class="st">&quot;NMDS&quot;</span>, <span class="at">colour_by =</span> <span class="st">&quot;shannon&quot;</span>)</span></code></pre></div>
<p><img src="19_visualization_techniques_files/figure-html/unnamed-chunk-7-1.png" width="672" /></p>
<p>Multiple combinations of coordinates or dimensions can also be integrated into a multi-panel arrangement.</p>
<div class="sourceCode" id="cb397"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb397-1"><a href="viz-chapter.html#cb397-1" tabindex="-1"></a><span class="co"># perform MDS coordination method</span></span>
<span id="cb397-2"><a href="viz-chapter.html#cb397-2" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">runMDS</span>(tse,</span>
<span id="cb397-3"><a href="viz-chapter.html#cb397-3" tabindex="-1"></a>              <span class="at">FUN =</span> vegan<span class="sc">::</span>vegdist,</span>
<span id="cb397-4"><a href="viz-chapter.html#cb397-4" tabindex="-1"></a>              <span class="at">method =</span> <span class="st">&quot;bray&quot;</span>,</span>
<span id="cb397-5"><a href="viz-chapter.html#cb397-5" tabindex="-1"></a>              <span class="at">name =</span> <span class="st">&quot;MDS&quot;</span>,</span>
<span id="cb397-6"><a href="viz-chapter.html#cb397-6" tabindex="-1"></a>              <span class="at">assay.type =</span> <span class="st">&quot;counts&quot;</span>,</span>
<span id="cb397-7"><a href="viz-chapter.html#cb397-7" tabindex="-1"></a>              <span class="at">ncomponents =</span> <span class="dv">3</span>)</span>
<span id="cb397-8"><a href="viz-chapter.html#cb397-8" tabindex="-1"></a><span class="co"># plot results of a 3-component MDS on tse,</span></span>
<span id="cb397-9"><a href="viz-chapter.html#cb397-9" tabindex="-1"></a><span class="co"># coloured-scaled by faith diversity index</span></span>
<span id="cb397-10"><a href="viz-chapter.html#cb397-10" tabindex="-1"></a><span class="fu">plotReducedDim</span>(tse, <span class="st">&quot;MDS&quot;</span>, <span class="at">ncomponents =</span> <span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>), <span class="at">colour_by =</span> <span class="st">&quot;faith&quot;</span>)</span></code></pre></div>
<p><img src="19_visualization_techniques_files/figure-html/unnamed-chunk-8-1.png" width="672" /></p>
<p>Similarly to iterating <code>plotColData</code> over indices of alpha diversity,
<code>lapply</code> can be used in combination with <em>patchwork</em> to recursively
apply <code>plotReducedDim</code> and visually compare results among various
coordination methods.</p>
<div class="sourceCode" id="cb398"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb398-1"><a href="viz-chapter.html#cb398-1" tabindex="-1"></a><span class="co"># generate plots for MDS and NMDS methods</span></span>
<span id="cb398-2"><a href="viz-chapter.html#cb398-2" tabindex="-1"></a><span class="co"># and store them into a list</span></span>
<span id="cb398-3"><a href="viz-chapter.html#cb398-3" tabindex="-1"></a>plots <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">c</span>(<span class="st">&quot;MDS&quot;</span>, <span class="st">&quot;NMDS&quot;</span>),</span>
<span id="cb398-4"><a href="viz-chapter.html#cb398-4" tabindex="-1"></a>                plotReducedDim,</span>
<span id="cb398-5"><a href="viz-chapter.html#cb398-5" tabindex="-1"></a>                <span class="at">object =</span> tse,</span>
<span id="cb398-6"><a href="viz-chapter.html#cb398-6" tabindex="-1"></a>                <span class="at">colour_by =</span> <span class="st">&quot;shannon&quot;</span>)</span>
<span id="cb398-7"><a href="viz-chapter.html#cb398-7" tabindex="-1"></a><span class="co"># combine plots with patchwork</span></span>
<span id="cb398-8"><a href="viz-chapter.html#cb398-8" tabindex="-1"></a>plots[[<span class="dv">1</span>]] <span class="sc">+</span> plots[[<span class="dv">2</span>]] <span class="sc">+</span></span>
<span id="cb398-9"><a href="viz-chapter.html#cb398-9" tabindex="-1"></a>  <span class="fu">plot_layout</span>(<span class="at">guides =</span> <span class="st">&quot;collect&quot;</span>)</span></code></pre></div>
<p><img src="19_visualization_techniques_files/figure-html/unnamed-chunk-9-1.png" width="672" /></p>
<p>For similar examples, readers are referred to chapter
<a href="community-similarity.html#community-similarity">8</a>. Further material on the graphic
capabilities of <em>patchwork</em> is available in its <a href="https://patchwork.data-imaginist.com/articles/patchwork.html">official package
tutorial</a>.</p>
</div>
</div>
<div id="statistical-analysis" class="section level2 hasAnchor" number="14.3">
<h2><span class="header-section-number">14.3</span> Statistical analysis<a href="viz-chapter.html#statistical-analysis" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="heatmaps" class="section level3 hasAnchor" number="14.3.1">
<h3><span class="header-section-number">14.3.1</span> Heatmaps<a href="viz-chapter.html#heatmaps" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>As described in chapter <a href="microbiome-community.html#visual-composition">9.1</a>, bar plots and
heatmaps can offer a useful insight into the composition of a
community. Simple methods involve the functions <code>plotAbundance</code> and
<code>geom_tile</code> in combination with <code>scale_fill_gradientn</code> from the
packages <em>miaViz</em> and <em>ggplot2</em>, respectively.</p>
<p>For instance, below the composition of multiple samples (x axis) is
reported in terms of relative abundances (y axis) for the top 10 taxa
at the Order rank. Bar plots and heatmaps with analogous information
at the Phylum level are available in the aforementioned chapter.</p>
<div class="sourceCode" id="cb399"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb399-1"><a href="viz-chapter.html#cb399-1" tabindex="-1"></a><span class="co"># agglomerate tse by Order</span></span>
<span id="cb399-2"><a href="viz-chapter.html#cb399-2" tabindex="-1"></a>tse_order <span class="ot">&lt;-</span> <span class="fu">mergeFeaturesByRank</span>(tse,</span>
<span id="cb399-3"><a href="viz-chapter.html#cb399-3" tabindex="-1"></a>                                <span class="at">rank =</span> <span class="st">&quot;Order&quot;</span>,</span>
<span id="cb399-4"><a href="viz-chapter.html#cb399-4" tabindex="-1"></a>                                <span class="at">onRankOnly =</span> <span class="cn">TRUE</span>)</span>
<span id="cb399-5"><a href="viz-chapter.html#cb399-5" tabindex="-1"></a><span class="co"># transform counts into relative abundance</span></span>
<span id="cb399-6"><a href="viz-chapter.html#cb399-6" tabindex="-1"></a>tse_order <span class="ot">&lt;-</span> <span class="fu">transformAssay</span>(tse_order,</span>
<span id="cb399-7"><a href="viz-chapter.html#cb399-7" tabindex="-1"></a>                              <span class="at">assay.type =</span> <span class="st">&quot;counts&quot;</span>,</span>
<span id="cb399-8"><a href="viz-chapter.html#cb399-8" tabindex="-1"></a>                              <span class="at">method =</span> <span class="st">&quot;relabundance&quot;</span>)</span>
<span id="cb399-9"><a href="viz-chapter.html#cb399-9" tabindex="-1"></a><span class="co"># get top orders</span></span>
<span id="cb399-10"><a href="viz-chapter.html#cb399-10" tabindex="-1"></a>top_taxa <span class="ot">&lt;-</span> <span class="fu">getTopFeatures</span>(tse_order,</span>
<span id="cb399-11"><a href="viz-chapter.html#cb399-11" tabindex="-1"></a>                       <span class="at">top =</span> <span class="dv">10</span>,</span>
<span id="cb399-12"><a href="viz-chapter.html#cb399-12" tabindex="-1"></a>                       <span class="at">assay.type =</span> <span class="st">&quot;relabundance&quot;</span>)</span>
<span id="cb399-13"><a href="viz-chapter.html#cb399-13" tabindex="-1"></a><span class="co"># leave only names for top 10 orders and label the rest with &quot;Other&quot;</span></span>
<span id="cb399-14"><a href="viz-chapter.html#cb399-14" tabindex="-1"></a>order_renamed <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">rowData</span>(tse_order)<span class="sc">$</span>Order,</span>
<span id="cb399-15"><a href="viz-chapter.html#cb399-15" tabindex="-1"></a>                   <span class="cf">function</span>(x){<span class="cf">if</span> (x <span class="sc">%in%</span> top_taxa) {x} <span class="cf">else</span> {<span class="st">&quot;Other&quot;</span>}})</span>
<span id="cb399-16"><a href="viz-chapter.html#cb399-16" tabindex="-1"></a><span class="fu">rowData</span>(tse_order)<span class="sc">$</span>Order <span class="ot">&lt;-</span> <span class="fu">as.character</span>(order_renamed)</span>
<span id="cb399-17"><a href="viz-chapter.html#cb399-17" tabindex="-1"></a><span class="co"># plot composition as a bar plot</span></span>
<span id="cb399-18"><a href="viz-chapter.html#cb399-18" tabindex="-1"></a><span class="fu">plotAbundance</span>(tse_order,</span>
<span id="cb399-19"><a href="viz-chapter.html#cb399-19" tabindex="-1"></a>              <span class="at">assay.type =</span> <span class="st">&quot;relabundance&quot;</span>,</span>
<span id="cb399-20"><a href="viz-chapter.html#cb399-20" tabindex="-1"></a>              <span class="at">rank =</span> <span class="st">&quot;Order&quot;</span>,</span>
<span id="cb399-21"><a href="viz-chapter.html#cb399-21" tabindex="-1"></a>              <span class="at">order_rank_by =</span> <span class="st">&quot;abund&quot;</span>,</span>
<span id="cb399-22"><a href="viz-chapter.html#cb399-22" tabindex="-1"></a>              <span class="at">order_sample_by =</span> <span class="st">&quot;Clostridiales&quot;</span>)</span></code></pre></div>
<p><img src="19_visualization_techniques_files/figure-html/plotAbundance1-1.png" width="672" /></p>
<p>To add a sample annotation, you can combine plots that you get from the output
of <em>plotAbundance</em>.</p>
<div class="sourceCode" id="cb400"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb400-1"><a href="viz-chapter.html#cb400-1" tabindex="-1"></a><span class="co"># Create plots</span></span>
<span id="cb400-2"><a href="viz-chapter.html#cb400-2" tabindex="-1"></a>plots <span class="ot">&lt;-</span> <span class="fu">plotAbundance</span>(tse_order,</span>
<span id="cb400-3"><a href="viz-chapter.html#cb400-3" tabindex="-1"></a>            <span class="at">assay.type =</span> <span class="st">&quot;relabundance&quot;</span>,</span>
<span id="cb400-4"><a href="viz-chapter.html#cb400-4" tabindex="-1"></a>        <span class="at">rank =</span> <span class="st">&quot;Order&quot;</span>,</span>
<span id="cb400-5"><a href="viz-chapter.html#cb400-5" tabindex="-1"></a>            <span class="at">order_rank_by =</span> <span class="st">&quot;abund&quot;</span>,</span>
<span id="cb400-6"><a href="viz-chapter.html#cb400-6" tabindex="-1"></a>        <span class="at">order_sample_by =</span> <span class="st">&quot;Clostridiales&quot;</span>,</span>
<span id="cb400-7"><a href="viz-chapter.html#cb400-7" tabindex="-1"></a>            <span class="at">features =</span> <span class="st">&quot;SampleType&quot;</span>)</span>
<span id="cb400-8"><a href="viz-chapter.html#cb400-8" tabindex="-1"></a></span>
<span id="cb400-9"><a href="viz-chapter.html#cb400-9" tabindex="-1"></a><span class="co"># Modify the legend of the first plot to be smaller </span></span>
<span id="cb400-10"><a href="viz-chapter.html#cb400-10" tabindex="-1"></a>plots[[<span class="dv">1</span>]] <span class="ot">&lt;-</span> plots[[<span class="dv">1</span>]] <span class="sc">+</span></span>
<span id="cb400-11"><a href="viz-chapter.html#cb400-11" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">legend.key.size =</span> <span class="fu">unit</span>(<span class="fl">0.3</span>, <span class="st">&#39;cm&#39;</span>),</span>
<span id="cb400-12"><a href="viz-chapter.html#cb400-12" tabindex="-1"></a>          <span class="at">legend.text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">6</span>),</span>
<span id="cb400-13"><a href="viz-chapter.html#cb400-13" tabindex="-1"></a>          <span class="at">legend.title =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">8</span>))</span>
<span id="cb400-14"><a href="viz-chapter.html#cb400-14" tabindex="-1"></a></span>
<span id="cb400-15"><a href="viz-chapter.html#cb400-15" tabindex="-1"></a><span class="co"># Modify the legend of the second plot to be smaller </span></span>
<span id="cb400-16"><a href="viz-chapter.html#cb400-16" tabindex="-1"></a>plots[[<span class="dv">2</span>]] <span class="ot">&lt;-</span> plots[[<span class="dv">2</span>]] <span class="sc">+</span></span>
<span id="cb400-17"><a href="viz-chapter.html#cb400-17" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">legend.key.height =</span> <span class="fu">unit</span>(<span class="fl">0.3</span>, <span class="st">&#39;cm&#39;</span>),</span>
<span id="cb400-18"><a href="viz-chapter.html#cb400-18" tabindex="-1"></a>          <span class="at">legend.key.width =</span> <span class="fu">unit</span>(<span class="fl">0.3</span>, <span class="st">&#39;cm&#39;</span>),</span>
<span id="cb400-19"><a href="viz-chapter.html#cb400-19" tabindex="-1"></a>          <span class="at">legend.text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">6</span>),</span>
<span id="cb400-20"><a href="viz-chapter.html#cb400-20" tabindex="-1"></a>          <span class="at">legend.title =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">8</span>),</span>
<span id="cb400-21"><a href="viz-chapter.html#cb400-21" tabindex="-1"></a>          <span class="at">legend.direction =</span> <span class="st">&quot;vertical&quot;</span>)</span>
<span id="cb400-22"><a href="viz-chapter.html#cb400-22" tabindex="-1"></a></span>
<span id="cb400-23"><a href="viz-chapter.html#cb400-23" tabindex="-1"></a><span class="co"># Load required packages</span></span>
<span id="cb400-24"><a href="viz-chapter.html#cb400-24" tabindex="-1"></a><span class="fu">library</span>(ggpubr)</span>
<span id="cb400-25"><a href="viz-chapter.html#cb400-25" tabindex="-1"></a><span class="fu">library</span>(patchwork) </span>
<span id="cb400-26"><a href="viz-chapter.html#cb400-26" tabindex="-1"></a><span class="co"># Combine legends</span></span>
<span id="cb400-27"><a href="viz-chapter.html#cb400-27" tabindex="-1"></a>legend <span class="ot">&lt;-</span> <span class="fu">wrap_plots</span>(<span class="fu">as_ggplot</span>(<span class="fu">get_legend</span>(plots[[<span class="dv">1</span>]])), <span class="fu">as_ggplot</span>(<span class="fu">get_legend</span>(plots[[<span class="dv">2</span>]])), <span class="at">ncol =</span> <span class="dv">1</span>) </span>
<span id="cb400-28"><a href="viz-chapter.html#cb400-28" tabindex="-1"></a></span>
<span id="cb400-29"><a href="viz-chapter.html#cb400-29" tabindex="-1"></a><span class="co"># Remove legends from the plots</span></span>
<span id="cb400-30"><a href="viz-chapter.html#cb400-30" tabindex="-1"></a>plots[[<span class="dv">1</span>]] <span class="ot">&lt;-</span> plots[[<span class="dv">1</span>]] <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">&quot;none&quot;</span>)</span>
<span id="cb400-31"><a href="viz-chapter.html#cb400-31" tabindex="-1"></a>plots[[<span class="dv">2</span>]] <span class="ot">&lt;-</span> plots[[<span class="dv">2</span>]] <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">&quot;none&quot;</span>, <span class="at">axis.title.x=</span><span class="fu">element_blank</span>()) </span>
<span id="cb400-32"><a href="viz-chapter.html#cb400-32" tabindex="-1"></a></span>
<span id="cb400-33"><a href="viz-chapter.html#cb400-33" tabindex="-1"></a><span class="co"># Combine plots</span></span>
<span id="cb400-34"><a href="viz-chapter.html#cb400-34" tabindex="-1"></a>plot <span class="ot">&lt;-</span> <span class="fu">wrap_plots</span>(plots[[<span class="dv">2</span>]], plots[[<span class="dv">1</span>]], <span class="at">ncol =</span> <span class="dv">1</span>, <span class="at">heights =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">10</span>))</span>
<span id="cb400-35"><a href="viz-chapter.html#cb400-35" tabindex="-1"></a><span class="co"># Combine the plot with the legend</span></span>
<span id="cb400-36"><a href="viz-chapter.html#cb400-36" tabindex="-1"></a><span class="fu">wrap_plots</span>(plot, legend, <span class="at">nrow =</span> <span class="dv">1</span>, <span class="at">widths =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">1</span>))</span></code></pre></div>
<p><img src="19_visualization_techniques_files/figure-html/plotAbundance2-1.png" width="672" /></p>
<p>For more sophisticated visualizations than those produced with <code>plotAbundance</code>
and <em>ggplot2</em>, the packages <em>pheatmap</em> and <em>sechm</em> provide methods to include
feature and sample clusters in a heatmap, along with further functionality.</p>
<div class="sourceCode" id="cb401"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb401-1"><a href="viz-chapter.html#cb401-1" tabindex="-1"></a><span class="co"># Agglomerate tse by phylum</span></span>
<span id="cb401-2"><a href="viz-chapter.html#cb401-2" tabindex="-1"></a>tse_phylum <span class="ot">&lt;-</span> <span class="fu">mergeFeaturesByRank</span>(tse,</span>
<span id="cb401-3"><a href="viz-chapter.html#cb401-3" tabindex="-1"></a>                                <span class="at">rank =</span> <span class="st">&quot;Phylum&quot;</span>,</span>
<span id="cb401-4"><a href="viz-chapter.html#cb401-4" tabindex="-1"></a>                                <span class="at">onRankOnly =</span> <span class="cn">TRUE</span>)</span>
<span id="cb401-5"><a href="viz-chapter.html#cb401-5" tabindex="-1"></a></span>
<span id="cb401-6"><a href="viz-chapter.html#cb401-6" tabindex="-1"></a><span class="co"># Add clr-transformation on samples</span></span>
<span id="cb401-7"><a href="viz-chapter.html#cb401-7" tabindex="-1"></a>tse_phylum <span class="ot">&lt;-</span> <span class="fu">transformAssay</span>(tse_phylum, <span class="at">MARGIN =</span> <span class="st">&quot;samples&quot;</span>, <span class="at">method =</span> <span class="st">&quot;clr&quot;</span>, <span class="at">assay.type =</span> <span class="st">&quot;counts&quot;</span>, <span class="at">pseudocount=</span><span class="dv">1</span>)</span>
<span id="cb401-8"><a href="viz-chapter.html#cb401-8" tabindex="-1"></a></span>
<span id="cb401-9"><a href="viz-chapter.html#cb401-9" tabindex="-1"></a><span class="co"># Add z-transformation on features (taxa)</span></span>
<span id="cb401-10"><a href="viz-chapter.html#cb401-10" tabindex="-1"></a>tse_phylum <span class="ot">&lt;-</span> <span class="fu">transformAssay</span>(tse_phylum, <span class="at">assay.type =</span> <span class="st">&quot;clr&quot;</span>,</span>
<span id="cb401-11"><a href="viz-chapter.html#cb401-11" tabindex="-1"></a>                              <span class="at">MARGIN =</span> <span class="st">&quot;features&quot;</span>, </span>
<span id="cb401-12"><a href="viz-chapter.html#cb401-12" tabindex="-1"></a>                              <span class="at">method =</span> <span class="st">&quot;z&quot;</span>, <span class="at">name =</span> <span class="st">&quot;clr_z&quot;</span>)</span>
<span id="cb401-13"><a href="viz-chapter.html#cb401-13" tabindex="-1"></a></span>
<span id="cb401-14"><a href="viz-chapter.html#cb401-14" tabindex="-1"></a><span class="co"># Take subset: only samples from feces, skin, or tongue</span></span>
<span id="cb401-15"><a href="viz-chapter.html#cb401-15" tabindex="-1"></a>tse_phylum_subset <span class="ot">&lt;-</span> tse_phylum[ , tse_phylum<span class="sc">$</span>SampleType <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">&quot;Feces&quot;</span>, <span class="st">&quot;Skin&quot;</span>, <span class="st">&quot;Tongue&quot;</span>) ]</span>
<span id="cb401-16"><a href="viz-chapter.html#cb401-16" tabindex="-1"></a></span>
<span id="cb401-17"><a href="viz-chapter.html#cb401-17" tabindex="-1"></a><span class="co"># Add clr-transformation</span></span>
<span id="cb401-18"><a href="viz-chapter.html#cb401-18" tabindex="-1"></a>tse_phylum_subset <span class="ot">&lt;-</span> <span class="fu">transformAssay</span>(tse_phylum_subset, <span class="at">method =</span> <span class="st">&quot;clr&quot;</span>,</span>
<span id="cb401-19"><a href="viz-chapter.html#cb401-19" tabindex="-1"></a>                                     <span class="at">MARGIN=</span><span class="st">&quot;samples&quot;</span>,</span>
<span id="cb401-20"><a href="viz-chapter.html#cb401-20" tabindex="-1"></a>                                     <span class="at">assay.type =</span> <span class="st">&quot;counts&quot;</span>, <span class="at">pseudocount=</span><span class="dv">1</span>)</span>
<span id="cb401-21"><a href="viz-chapter.html#cb401-21" tabindex="-1"></a><span class="co"># Does z-transformation</span></span>
<span id="cb401-22"><a href="viz-chapter.html#cb401-22" tabindex="-1"></a>tse_phylum_subset <span class="ot">&lt;-</span> <span class="fu">transformAssay</span>(tse_phylum_subset, <span class="at">assay.type =</span> <span class="st">&quot;clr&quot;</span>,</span>
<span id="cb401-23"><a href="viz-chapter.html#cb401-23" tabindex="-1"></a>                                     <span class="at">MARGIN =</span> <span class="st">&quot;features&quot;</span>, </span>
<span id="cb401-24"><a href="viz-chapter.html#cb401-24" tabindex="-1"></a>                                     <span class="at">method =</span> <span class="st">&quot;z&quot;</span>, <span class="at">name =</span> <span class="st">&quot;clr_z&quot;</span>)</span>
<span id="cb401-25"><a href="viz-chapter.html#cb401-25" tabindex="-1"></a></span>
<span id="cb401-26"><a href="viz-chapter.html#cb401-26" tabindex="-1"></a><span class="co"># Get n most abundant taxa, and subsets the data by them</span></span>
<span id="cb401-27"><a href="viz-chapter.html#cb401-27" tabindex="-1"></a>top_taxa <span class="ot">&lt;-</span> <span class="fu">getTopFeatures</span>(tse_phylum_subset, <span class="at">top =</span> <span class="dv">20</span>)</span>
<span id="cb401-28"><a href="viz-chapter.html#cb401-28" tabindex="-1"></a>tse_phylum_subset <span class="ot">&lt;-</span> tse_phylum_subset[top_taxa, ]</span>
<span id="cb401-29"><a href="viz-chapter.html#cb401-29" tabindex="-1"></a></span>
<span id="cb401-30"><a href="viz-chapter.html#cb401-30" tabindex="-1"></a><span class="co"># Gets the assay table</span></span>
<span id="cb401-31"><a href="viz-chapter.html#cb401-31" tabindex="-1"></a>mat <span class="ot">&lt;-</span> <span class="fu">assay</span>(tse_phylum_subset, <span class="st">&quot;clr_z&quot;</span>)</span>
<span id="cb401-32"><a href="viz-chapter.html#cb401-32" tabindex="-1"></a></span>
<span id="cb401-33"><a href="viz-chapter.html#cb401-33" tabindex="-1"></a><span class="co"># Creates the heatmap</span></span>
<span id="cb401-34"><a href="viz-chapter.html#cb401-34" tabindex="-1"></a><span class="fu">pheatmap</span>(mat)</span></code></pre></div>
<p><img src="19_visualization_techniques_files/figure-html/pheatmap1-1.png" width="672" /></p>
<p>We can cluster both samples and features hierarchically and add them to the
x and y axes of the heatmap, respectively.</p>
<div class="sourceCode" id="cb402"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb402-1"><a href="viz-chapter.html#cb402-1" tabindex="-1"></a><span class="co"># Hierarchical clustering</span></span>
<span id="cb402-2"><a href="viz-chapter.html#cb402-2" tabindex="-1"></a>taxa_hclust <span class="ot">&lt;-</span> <span class="fu">hclust</span>(<span class="fu">dist</span>(mat), <span class="at">method =</span> <span class="st">&quot;complete&quot;</span>)</span>
<span id="cb402-3"><a href="viz-chapter.html#cb402-3" tabindex="-1"></a></span>
<span id="cb402-4"><a href="viz-chapter.html#cb402-4" tabindex="-1"></a><span class="co"># Creates a phylogenetic tree</span></span>
<span id="cb402-5"><a href="viz-chapter.html#cb402-5" tabindex="-1"></a>taxa_tree <span class="ot">&lt;-</span> <span class="fu">as.phylo</span>(taxa_hclust)</span>
<span id="cb402-6"><a href="viz-chapter.html#cb402-6" tabindex="-1"></a></span>
<span id="cb402-7"><a href="viz-chapter.html#cb402-7" tabindex="-1"></a><span class="co"># Plot taxa tree</span></span>
<span id="cb402-8"><a href="viz-chapter.html#cb402-8" tabindex="-1"></a>taxa_tree <span class="ot">&lt;-</span> <span class="fu">ggtree</span>(taxa_tree) <span class="sc">+</span> </span>
<span id="cb402-9"><a href="viz-chapter.html#cb402-9" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.margin=</span><span class="fu">margin</span>(<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>)) <span class="co"># removes margins</span></span>
<span id="cb402-10"><a href="viz-chapter.html#cb402-10" tabindex="-1"></a></span>
<span id="cb402-11"><a href="viz-chapter.html#cb402-11" tabindex="-1"></a><span class="co"># Get order of taxa in plot</span></span>
<span id="cb402-12"><a href="viz-chapter.html#cb402-12" tabindex="-1"></a>taxa_ordered <span class="ot">&lt;-</span> <span class="fu">get_taxa_name</span>(taxa_tree)</span>
<span id="cb402-13"><a href="viz-chapter.html#cb402-13" tabindex="-1"></a></span>
<span id="cb402-14"><a href="viz-chapter.html#cb402-14" tabindex="-1"></a><span class="co"># to view the tree, run</span></span>
<span id="cb402-15"><a href="viz-chapter.html#cb402-15" tabindex="-1"></a><span class="co"># taxa_tree</span></span></code></pre></div>
<p>Based on phylo tree, we decide to create three clusters.</p>
<div class="sourceCode" id="cb403"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb403-1"><a href="viz-chapter.html#cb403-1" tabindex="-1"></a><span class="co"># Creates clusters</span></span>
<span id="cb403-2"><a href="viz-chapter.html#cb403-2" tabindex="-1"></a>taxa_clusters <span class="ot">&lt;-</span> <span class="fu">cutree</span>(<span class="at">tree =</span> taxa_hclust, <span class="at">k =</span> <span class="dv">3</span>)</span>
<span id="cb403-3"><a href="viz-chapter.html#cb403-3" tabindex="-1"></a></span>
<span id="cb403-4"><a href="viz-chapter.html#cb403-4" tabindex="-1"></a><span class="co"># Converts into data frame</span></span>
<span id="cb403-5"><a href="viz-chapter.html#cb403-5" tabindex="-1"></a>taxa_clusters <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">clusters =</span> taxa_clusters)</span>
<span id="cb403-6"><a href="viz-chapter.html#cb403-6" tabindex="-1"></a>taxa_clusters<span class="sc">$</span>clusters <span class="ot">&lt;-</span> <span class="fu">factor</span>(taxa_clusters<span class="sc">$</span>clusters)</span>
<span id="cb403-7"><a href="viz-chapter.html#cb403-7" tabindex="-1"></a></span>
<span id="cb403-8"><a href="viz-chapter.html#cb403-8" tabindex="-1"></a><span class="co"># Order data so that it&#39;s same as in phylo tree</span></span>
<span id="cb403-9"><a href="viz-chapter.html#cb403-9" tabindex="-1"></a>taxa_clusters <span class="ot">&lt;-</span> taxa_clusters[taxa_ordered, , drop <span class="ot">=</span> <span class="cn">FALSE</span>] </span>
<span id="cb403-10"><a href="viz-chapter.html#cb403-10" tabindex="-1"></a></span>
<span id="cb403-11"><a href="viz-chapter.html#cb403-11" tabindex="-1"></a><span class="co"># Prints taxa and their clusters</span></span>
<span id="cb403-12"><a href="viz-chapter.html#cb403-12" tabindex="-1"></a>taxa_clusters</span></code></pre></div>
<pre><code>##                  clusters
## Chloroflexi             3
## Actinobacteria          3
## Crenarchaeota           3
## Planctomycetes          3
## Gemmatimonadetes        3
## Thermi                  3
## Acidobacteria           3
## Spirochaetes            2
## Fusobacteria            2
## SR1                     2
## Cyanobacteria           2
## Proteobacteria          2
## Synergistetes           2
## Lentisphaerae           1
## Bacteroidetes           1
## Verrucomicrobia         1
## Tenericutes             1
## Firmicutes              1
## Euryarchaeota           1
## SAR406                  1</code></pre>
<p>The information on the clusters is then added to the feature meta data.</p>
<div class="sourceCode" id="cb405"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb405-1"><a href="viz-chapter.html#cb405-1" tabindex="-1"></a><span class="co"># Adds information to rowData</span></span>
<span id="cb405-2"><a href="viz-chapter.html#cb405-2" tabindex="-1"></a><span class="fu">rowData</span>(tse_phylum_subset)<span class="sc">$</span>clusters <span class="ot">&lt;-</span> taxa_clusters[<span class="fu">order</span>(<span class="fu">match</span>(<span class="fu">rownames</span>(taxa_clusters), <span class="fu">rownames</span>(tse_phylum_subset))), ]</span>
<span id="cb405-3"><a href="viz-chapter.html#cb405-3" tabindex="-1"></a></span>
<span id="cb405-4"><a href="viz-chapter.html#cb405-4" tabindex="-1"></a><span class="co"># Prints taxa and their clusters</span></span>
<span id="cb405-5"><a href="viz-chapter.html#cb405-5" tabindex="-1"></a><span class="fu">rowData</span>(tse_phylum_subset)<span class="sc">$</span>clusters</span></code></pre></div>
<pre><code>##  [1] 1 1 2 3 2 2 1 1 1 1 3 2 3 3 3 2 2 3 3 1
## Levels: 1 2 3</code></pre>
<p>Similarly, samples are hierarchically grouped into clusters, the most suitable
number of clusters for the plot is selected and the new information is stored
into the sample meta data.</p>
<div class="sourceCode" id="cb407"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb407-1"><a href="viz-chapter.html#cb407-1" tabindex="-1"></a><span class="co"># Hierarchical clustering</span></span>
<span id="cb407-2"><a href="viz-chapter.html#cb407-2" tabindex="-1"></a>sample_hclust <span class="ot">&lt;-</span> <span class="fu">hclust</span>(<span class="fu">dist</span>(<span class="fu">t</span>(mat)), <span class="at">method =</span> <span class="st">&quot;complete&quot;</span>)</span>
<span id="cb407-3"><a href="viz-chapter.html#cb407-3" tabindex="-1"></a></span>
<span id="cb407-4"><a href="viz-chapter.html#cb407-4" tabindex="-1"></a><span class="co"># Creates a phylogenetic tree</span></span>
<span id="cb407-5"><a href="viz-chapter.html#cb407-5" tabindex="-1"></a>sample_tree <span class="ot">&lt;-</span> <span class="fu">as.phylo</span>(sample_hclust)</span>
<span id="cb407-6"><a href="viz-chapter.html#cb407-6" tabindex="-1"></a></span>
<span id="cb407-7"><a href="viz-chapter.html#cb407-7" tabindex="-1"></a><span class="co"># Plot sample tree</span></span>
<span id="cb407-8"><a href="viz-chapter.html#cb407-8" tabindex="-1"></a>sample_tree <span class="ot">&lt;-</span> <span class="fu">ggtree</span>(sample_tree) <span class="sc">+</span> <span class="fu">layout_dendrogram</span>() <span class="sc">+</span> </span>
<span id="cb407-9"><a href="viz-chapter.html#cb407-9" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.margin=</span><span class="fu">margin</span>(<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>)) <span class="co"># removes margins</span></span>
<span id="cb407-10"><a href="viz-chapter.html#cb407-10" tabindex="-1"></a></span>
<span id="cb407-11"><a href="viz-chapter.html#cb407-11" tabindex="-1"></a><span class="co"># Get order of samples in plot</span></span>
<span id="cb407-12"><a href="viz-chapter.html#cb407-12" tabindex="-1"></a>samples_ordered <span class="ot">&lt;-</span> <span class="fu">rev</span>(<span class="fu">get_taxa_name</span>(sample_tree))</span>
<span id="cb407-13"><a href="viz-chapter.html#cb407-13" tabindex="-1"></a></span>
<span id="cb407-14"><a href="viz-chapter.html#cb407-14" tabindex="-1"></a><span class="co"># to view the tree, run</span></span>
<span id="cb407-15"><a href="viz-chapter.html#cb407-15" tabindex="-1"></a><span class="co"># sample_tree</span></span>
<span id="cb407-16"><a href="viz-chapter.html#cb407-16" tabindex="-1"></a></span>
<span id="cb407-17"><a href="viz-chapter.html#cb407-17" tabindex="-1"></a><span class="co"># Creates clusters</span></span>
<span id="cb407-18"><a href="viz-chapter.html#cb407-18" tabindex="-1"></a>sample_clusters <span class="ot">&lt;-</span> <span class="fu">factor</span>(<span class="fu">cutree</span>(<span class="at">tree =</span> sample_hclust, <span class="at">k =</span> <span class="dv">3</span>))</span>
<span id="cb407-19"><a href="viz-chapter.html#cb407-19" tabindex="-1"></a></span>
<span id="cb407-20"><a href="viz-chapter.html#cb407-20" tabindex="-1"></a><span class="co"># Converts into data frame</span></span>
<span id="cb407-21"><a href="viz-chapter.html#cb407-21" tabindex="-1"></a>sample_data <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">clusters =</span> sample_clusters)</span>
<span id="cb407-22"><a href="viz-chapter.html#cb407-22" tabindex="-1"></a></span>
<span id="cb407-23"><a href="viz-chapter.html#cb407-23" tabindex="-1"></a><span class="co"># Order data so that it&#39;s same as in phylo tree</span></span>
<span id="cb407-24"><a href="viz-chapter.html#cb407-24" tabindex="-1"></a>sample_data <span class="ot">&lt;-</span> sample_data[samples_ordered, , drop <span class="ot">=</span> <span class="cn">FALSE</span>] </span>
<span id="cb407-25"><a href="viz-chapter.html#cb407-25" tabindex="-1"></a></span>
<span id="cb407-26"><a href="viz-chapter.html#cb407-26" tabindex="-1"></a><span class="co"># Order data based on </span></span>
<span id="cb407-27"><a href="viz-chapter.html#cb407-27" tabindex="-1"></a>tse_phylum_subset <span class="ot">&lt;-</span> tse_phylum_subset[ , <span class="fu">rownames</span>(sample_data)]</span>
<span id="cb407-28"><a href="viz-chapter.html#cb407-28" tabindex="-1"></a></span>
<span id="cb407-29"><a href="viz-chapter.html#cb407-29" tabindex="-1"></a><span class="co"># Add sample type data</span></span>
<span id="cb407-30"><a href="viz-chapter.html#cb407-30" tabindex="-1"></a>sample_data<span class="sc">$</span>sample_types <span class="ot">&lt;-</span> <span class="fu">unfactor</span>(<span class="fu">colData</span>(tse_phylum_subset)<span class="sc">$</span>SampleType)</span>
<span id="cb407-31"><a href="viz-chapter.html#cb407-31" tabindex="-1"></a></span>
<span id="cb407-32"><a href="viz-chapter.html#cb407-32" tabindex="-1"></a>sample_data</span></code></pre></div>
<pre><code>##         clusters sample_types
## M11Plmr        2         Skin
## M31Plmr        2         Skin
## F21Plmr        2         Skin
## M31Fcsw        1        Feces
## M11Fcsw        1        Feces
## TS28           3        Feces
## TS29           3        Feces
## M31Tong        3       Tongue
## M11Tong        3       Tongue</code></pre>
<p>Now we can create heatmap with additional annotations.</p>
<div class="sourceCode" id="cb409"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb409-1"><a href="viz-chapter.html#cb409-1" tabindex="-1"></a><span class="co"># Determines the scaling of colorss</span></span>
<span id="cb409-2"><a href="viz-chapter.html#cb409-2" tabindex="-1"></a><span class="co"># Scale colors</span></span>
<span id="cb409-3"><a href="viz-chapter.html#cb409-3" tabindex="-1"></a>breaks <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="sc">-</span><span class="fu">ceiling</span>(<span class="fu">max</span>(<span class="fu">abs</span>(mat))), <span class="fu">ceiling</span>(<span class="fu">max</span>(<span class="fu">abs</span>(mat))), </span>
<span id="cb409-4"><a href="viz-chapter.html#cb409-4" tabindex="-1"></a>              <span class="at">length.out =</span> <span class="fu">ifelse</span>( <span class="fu">max</span>(<span class="fu">abs</span>(mat))<span class="sc">&gt;</span><span class="dv">5</span>, <span class="dv">2</span><span class="sc">*</span><span class="fu">ceiling</span>(<span class="fu">max</span>(<span class="fu">abs</span>(mat))), <span class="dv">10</span> ) )</span>
<span id="cb409-5"><a href="viz-chapter.html#cb409-5" tabindex="-1"></a>colors <span class="ot">&lt;-</span> <span class="fu">colorRampPalette</span>(<span class="fu">c</span>(<span class="st">&quot;darkblue&quot;</span>, <span class="st">&quot;blue&quot;</span>, <span class="st">&quot;white&quot;</span>, <span class="st">&quot;red&quot;</span>, <span class="st">&quot;darkred&quot;</span>))(<span class="fu">length</span>(breaks)<span class="sc">-</span><span class="dv">1</span>)</span>
<span id="cb409-6"><a href="viz-chapter.html#cb409-6" tabindex="-1"></a></span>
<span id="cb409-7"><a href="viz-chapter.html#cb409-7" tabindex="-1"></a><span class="fu">pheatmap</span>(mat, <span class="at">annotation_row =</span> taxa_clusters, </span>
<span id="cb409-8"><a href="viz-chapter.html#cb409-8" tabindex="-1"></a>         <span class="at">annotation_col =</span> sample_data,</span>
<span id="cb409-9"><a href="viz-chapter.html#cb409-9" tabindex="-1"></a>         <span class="at">breaks =</span> breaks,</span>
<span id="cb409-10"><a href="viz-chapter.html#cb409-10" tabindex="-1"></a>         <span class="at">color =</span> colors)</span></code></pre></div>
<p><img src="19_visualization_techniques_files/figure-html/pheatmap6-1.png" width="672" /></p>
<p>The package <em>sechm</em> allows for further visual capabilities and flexibility.
In this case, the clustering step is automatically performed by the plotting
function and does not need to be executed in advance.</p>
<div class="sourceCode" id="cb410"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb410-1"><a href="viz-chapter.html#cb410-1" tabindex="-1"></a><span class="co"># Stores annotation colros to metadata</span></span>
<span id="cb410-2"><a href="viz-chapter.html#cb410-2" tabindex="-1"></a><span class="fu">metadata</span>(tse_phylum_subset)<span class="sc">$</span>anno_colors<span class="sc">$</span>SampleType <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="at">Feces =</span> <span class="st">&quot;blue&quot;</span>, </span>
<span id="cb410-3"><a href="viz-chapter.html#cb410-3" tabindex="-1"></a>                                                        <span class="at">Skin =</span> <span class="st">&quot;red&quot;</span>, </span>
<span id="cb410-4"><a href="viz-chapter.html#cb410-4" tabindex="-1"></a>                                                        <span class="at">Tongue =</span> <span class="st">&quot;gray&quot;</span>)</span>
<span id="cb410-5"><a href="viz-chapter.html#cb410-5" tabindex="-1"></a></span>
<span id="cb410-6"><a href="viz-chapter.html#cb410-6" tabindex="-1"></a><span class="co"># Create a plot</span></span>
<span id="cb410-7"><a href="viz-chapter.html#cb410-7" tabindex="-1"></a><span class="fu">sechm</span>(tse_phylum_subset, </span>
<span id="cb410-8"><a href="viz-chapter.html#cb410-8" tabindex="-1"></a>      <span class="at">features =</span> <span class="fu">rownames</span>(tse_phylum_subset), </span>
<span id="cb410-9"><a href="viz-chapter.html#cb410-9" tabindex="-1"></a>      <span class="at">assayName =</span> <span class="st">&quot;clr&quot;</span>, </span>
<span id="cb410-10"><a href="viz-chapter.html#cb410-10" tabindex="-1"></a>      <span class="at">do.scale =</span> <span class="cn">TRUE</span>, </span>
<span id="cb410-11"><a href="viz-chapter.html#cb410-11" tabindex="-1"></a>      <span class="at">top_annotation =</span> <span class="fu">c</span>(<span class="st">&quot;SampleType&quot;</span>), </span>
<span id="cb410-12"><a href="viz-chapter.html#cb410-12" tabindex="-1"></a>      <span class="at">gaps_at =</span> <span class="st">&quot;SampleType&quot;</span>,</span>
<span id="cb410-13"><a href="viz-chapter.html#cb410-13" tabindex="-1"></a>      <span class="at">cluster_cols =</span> <span class="cn">TRUE</span>, <span class="at">cluster_rows =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p><img src="19_visualization_techniques_files/figure-html/sechm-1.png" width="672" /></p>
<p>It is also possible to create an analogous heatmap by just using the
<em>ggplot2</em> package. However, a relatively long code is required to
generate an identical output.</p>
<div class="sourceCode" id="cb411"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb411-1"><a href="viz-chapter.html#cb411-1" tabindex="-1"></a><span class="co"># Add feature names to column as a factor</span></span>
<span id="cb411-2"><a href="viz-chapter.html#cb411-2" tabindex="-1"></a>taxa_clusters<span class="sc">$</span>Feature <span class="ot">&lt;-</span> <span class="fu">rownames</span>(taxa_clusters)</span>
<span id="cb411-3"><a href="viz-chapter.html#cb411-3" tabindex="-1"></a>taxa_clusters<span class="sc">$</span>Feature <span class="ot">&lt;-</span> <span class="fu">factor</span>(taxa_clusters<span class="sc">$</span>Feature, <span class="at">levels =</span> taxa_clusters<span class="sc">$</span>Feature)</span>
<span id="cb411-4"><a href="viz-chapter.html#cb411-4" tabindex="-1"></a></span>
<span id="cb411-5"><a href="viz-chapter.html#cb411-5" tabindex="-1"></a><span class="co"># Create annotation plot</span></span>
<span id="cb411-6"><a href="viz-chapter.html#cb411-6" tabindex="-1"></a>row_annotation <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(taxa_clusters) <span class="sc">+</span> </span>
<span id="cb411-7"><a href="viz-chapter.html#cb411-7" tabindex="-1"></a>  <span class="fu">geom_tile</span>(<span class="fu">aes</span>(<span class="at">x =</span> <span class="cn">NA</span>, <span class="at">y =</span> Feature, <span class="at">fill =</span> clusters)) <span class="sc">+</span></span>
<span id="cb411-8"><a href="viz-chapter.html#cb411-8" tabindex="-1"></a>  <span class="fu">coord_equal</span>(<span class="at">ratio =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb411-9"><a href="viz-chapter.html#cb411-9" tabindex="-1"></a>  <span class="fu">theme</span>(</span>
<span id="cb411-10"><a href="viz-chapter.html#cb411-10" tabindex="-1"></a>        <span class="at">axis.text.x=</span><span class="fu">element_blank</span>(),</span>
<span id="cb411-11"><a href="viz-chapter.html#cb411-11" tabindex="-1"></a>        <span class="at">axis.text.y=</span><span class="fu">element_blank</span>(),</span>
<span id="cb411-12"><a href="viz-chapter.html#cb411-12" tabindex="-1"></a>        <span class="at">axis.ticks.y=</span><span class="fu">element_blank</span>(),</span>
<span id="cb411-13"><a href="viz-chapter.html#cb411-13" tabindex="-1"></a>        <span class="at">axis.title.y=</span><span class="fu">element_blank</span>(),</span>
<span id="cb411-14"><a href="viz-chapter.html#cb411-14" tabindex="-1"></a>        <span class="at">axis.title.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">90</span>, <span class="at">vjust =</span> <span class="fl">0.5</span>, <span class="at">hjust=</span><span class="dv">1</span>),</span>
<span id="cb411-15"><a href="viz-chapter.html#cb411-15" tabindex="-1"></a>        <span class="at">plot.margin=</span><span class="fu">margin</span>(<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>),</span>
<span id="cb411-16"><a href="viz-chapter.html#cb411-16" tabindex="-1"></a>        ) <span class="sc">+</span></span>
<span id="cb411-17"><a href="viz-chapter.html#cb411-17" tabindex="-1"></a>      <span class="fu">labs</span>(<span class="at">fill =</span> <span class="st">&quot;Clusters&quot;</span>, <span class="at">x =</span> <span class="st">&quot;Clusters&quot;</span>)</span>
<span id="cb411-18"><a href="viz-chapter.html#cb411-18" tabindex="-1"></a></span>
<span id="cb411-19"><a href="viz-chapter.html#cb411-19" tabindex="-1"></a><span class="co"># to view the notation, run</span></span>
<span id="cb411-20"><a href="viz-chapter.html#cb411-20" tabindex="-1"></a><span class="co"># row_annotation</span></span>
<span id="cb411-21"><a href="viz-chapter.html#cb411-21" tabindex="-1"></a></span>
<span id="cb411-22"><a href="viz-chapter.html#cb411-22" tabindex="-1"></a><span class="co"># Add sample names to one of the columns</span></span>
<span id="cb411-23"><a href="viz-chapter.html#cb411-23" tabindex="-1"></a>sample_data<span class="sc">$</span>sample <span class="ot">&lt;-</span> <span class="fu">factor</span>(<span class="fu">rownames</span>(sample_data), <span class="at">levels =</span> <span class="fu">rownames</span>(sample_data))</span>
<span id="cb411-24"><a href="viz-chapter.html#cb411-24" tabindex="-1"></a></span>
<span id="cb411-25"><a href="viz-chapter.html#cb411-25" tabindex="-1"></a><span class="co"># Create annotation plot</span></span>
<span id="cb411-26"><a href="viz-chapter.html#cb411-26" tabindex="-1"></a>sample_types_annotation <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(sample_data) <span class="sc">+</span></span>
<span id="cb411-27"><a href="viz-chapter.html#cb411-27" tabindex="-1"></a>  <span class="fu">scale_y_discrete</span>(<span class="at">position =</span> <span class="st">&quot;right&quot;</span>, <span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb411-28"><a href="viz-chapter.html#cb411-28" tabindex="-1"></a>  <span class="fu">geom_tile</span>(<span class="fu">aes</span>(<span class="at">y =</span> <span class="cn">NA</span>, <span class="at">x =</span> sample, <span class="at">fill =</span> sample_types)) <span class="sc">+</span></span>
<span id="cb411-29"><a href="viz-chapter.html#cb411-29" tabindex="-1"></a>  <span class="fu">coord_equal</span>(<span class="at">ratio =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb411-30"><a href="viz-chapter.html#cb411-30" tabindex="-1"></a>  <span class="fu">theme</span>(</span>
<span id="cb411-31"><a href="viz-chapter.html#cb411-31" tabindex="-1"></a>        <span class="at">axis.text.x=</span><span class="fu">element_blank</span>(),</span>
<span id="cb411-32"><a href="viz-chapter.html#cb411-32" tabindex="-1"></a>        <span class="at">axis.text.y=</span><span class="fu">element_blank</span>(),</span>
<span id="cb411-33"><a href="viz-chapter.html#cb411-33" tabindex="-1"></a>        <span class="at">axis.title.x=</span><span class="fu">element_blank</span>(),</span>
<span id="cb411-34"><a href="viz-chapter.html#cb411-34" tabindex="-1"></a>        <span class="at">axis.ticks.x=</span><span class="fu">element_blank</span>(),</span>
<span id="cb411-35"><a href="viz-chapter.html#cb411-35" tabindex="-1"></a>        <span class="at">plot.margin=</span><span class="fu">margin</span>(<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>),</span>
<span id="cb411-36"><a href="viz-chapter.html#cb411-36" tabindex="-1"></a>        <span class="at">axis.title.y.right =</span> <span class="fu">element_text</span>(<span class="at">angle=</span><span class="dv">0</span>, <span class="at">vjust =</span> <span class="fl">0.5</span>)</span>
<span id="cb411-37"><a href="viz-chapter.html#cb411-37" tabindex="-1"></a>        ) <span class="sc">+</span></span>
<span id="cb411-38"><a href="viz-chapter.html#cb411-38" tabindex="-1"></a>      <span class="fu">labs</span>(<span class="at">fill =</span> <span class="st">&quot;Sample types&quot;</span>, <span class="at">y =</span> <span class="st">&quot;Sample types&quot;</span>)</span>
<span id="cb411-39"><a href="viz-chapter.html#cb411-39" tabindex="-1"></a><span class="co"># to view the notation, run</span></span>
<span id="cb411-40"><a href="viz-chapter.html#cb411-40" tabindex="-1"></a><span class="co"># sample_types_annotation</span></span>
<span id="cb411-41"><a href="viz-chapter.html#cb411-41" tabindex="-1"></a></span>
<span id="cb411-42"><a href="viz-chapter.html#cb411-42" tabindex="-1"></a><span class="co"># Create annotation plot</span></span>
<span id="cb411-43"><a href="viz-chapter.html#cb411-43" tabindex="-1"></a>sample_clusters_annotation <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(sample_data) <span class="sc">+</span></span>
<span id="cb411-44"><a href="viz-chapter.html#cb411-44" tabindex="-1"></a>  <span class="fu">scale_y_discrete</span>(<span class="at">position =</span> <span class="st">&quot;right&quot;</span>, <span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb411-45"><a href="viz-chapter.html#cb411-45" tabindex="-1"></a>  <span class="fu">geom_tile</span>(<span class="fu">aes</span>(<span class="at">y =</span> <span class="cn">NA</span>, <span class="at">x =</span> sample, <span class="at">fill =</span> clusters)) <span class="sc">+</span></span>
<span id="cb411-46"><a href="viz-chapter.html#cb411-46" tabindex="-1"></a>  <span class="fu">coord_equal</span>(<span class="at">ratio =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb411-47"><a href="viz-chapter.html#cb411-47" tabindex="-1"></a>  <span class="fu">theme</span>(</span>
<span id="cb411-48"><a href="viz-chapter.html#cb411-48" tabindex="-1"></a>        <span class="at">axis.text.x=</span><span class="fu">element_blank</span>(),</span>
<span id="cb411-49"><a href="viz-chapter.html#cb411-49" tabindex="-1"></a>        <span class="at">axis.text.y=</span><span class="fu">element_blank</span>(),</span>
<span id="cb411-50"><a href="viz-chapter.html#cb411-50" tabindex="-1"></a>        <span class="at">axis.title.x=</span><span class="fu">element_blank</span>(),</span>
<span id="cb411-51"><a href="viz-chapter.html#cb411-51" tabindex="-1"></a>        <span class="at">axis.ticks.x=</span><span class="fu">element_blank</span>(),</span>
<span id="cb411-52"><a href="viz-chapter.html#cb411-52" tabindex="-1"></a>        <span class="at">plot.margin=</span><span class="fu">margin</span>(<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>),</span>
<span id="cb411-53"><a href="viz-chapter.html#cb411-53" tabindex="-1"></a>        <span class="at">axis.title.y.right =</span> <span class="fu">element_text</span>(<span class="at">angle=</span><span class="dv">0</span>, <span class="at">vjust =</span> <span class="fl">0.5</span>)</span>
<span id="cb411-54"><a href="viz-chapter.html#cb411-54" tabindex="-1"></a>        ) <span class="sc">+</span></span>
<span id="cb411-55"><a href="viz-chapter.html#cb411-55" tabindex="-1"></a>      <span class="fu">labs</span>(<span class="at">fill =</span> <span class="st">&quot;Clusters&quot;</span>, <span class="at">y =</span> <span class="st">&quot;Clusters&quot;</span>)</span>
<span id="cb411-56"><a href="viz-chapter.html#cb411-56" tabindex="-1"></a><span class="co"># to view the notation, run</span></span>
<span id="cb411-57"><a href="viz-chapter.html#cb411-57" tabindex="-1"></a><span class="co"># sample_clusters_annotation</span></span>
<span id="cb411-58"><a href="viz-chapter.html#cb411-58" tabindex="-1"></a></span>
<span id="cb411-59"><a href="viz-chapter.html#cb411-59" tabindex="-1"></a><span class="co"># Order data based on clusters and sample types</span></span>
<span id="cb411-60"><a href="viz-chapter.html#cb411-60" tabindex="-1"></a>mat <span class="ot">&lt;-</span> mat[<span class="fu">unfactor</span>(taxa_clusters<span class="sc">$</span>Feature), <span class="fu">unfactor</span>(sample_data<span class="sc">$</span>sample)]</span>
<span id="cb411-61"><a href="viz-chapter.html#cb411-61" tabindex="-1"></a></span>
<span id="cb411-62"><a href="viz-chapter.html#cb411-62" tabindex="-1"></a><span class="co"># ggplot requires data in melted format</span></span>
<span id="cb411-63"><a href="viz-chapter.html#cb411-63" tabindex="-1"></a>melted_mat <span class="ot">&lt;-</span> <span class="fu">melt</span>(mat)</span>
<span id="cb411-64"><a href="viz-chapter.html#cb411-64" tabindex="-1"></a><span class="fu">colnames</span>(melted_mat) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;Taxa&quot;</span>, <span class="st">&quot;Sample&quot;</span>, <span class="st">&quot;clr_z&quot;</span>)</span>
<span id="cb411-65"><a href="viz-chapter.html#cb411-65" tabindex="-1"></a></span>
<span id="cb411-66"><a href="viz-chapter.html#cb411-66" tabindex="-1"></a><span class="co"># Determines the scaling of colorss</span></span>
<span id="cb411-67"><a href="viz-chapter.html#cb411-67" tabindex="-1"></a>maxval <span class="ot">&lt;-</span> <span class="fu">round</span>(<span class="fu">max</span>(<span class="fu">abs</span>(melted_mat<span class="sc">$</span>clr_z)))</span>
<span id="cb411-68"><a href="viz-chapter.html#cb411-68" tabindex="-1"></a>limits <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="sc">-</span>maxval, maxval)</span>
<span id="cb411-69"><a href="viz-chapter.html#cb411-69" tabindex="-1"></a>breaks <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="at">from =</span> <span class="fu">min</span>(limits), <span class="at">to =</span> <span class="fu">max</span>(limits), <span class="at">by =</span> <span class="fl">0.5</span>)</span>
<span id="cb411-70"><a href="viz-chapter.html#cb411-70" tabindex="-1"></a>colours <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;darkblue&quot;</span>, <span class="st">&quot;blue&quot;</span>, <span class="st">&quot;white&quot;</span>, <span class="st">&quot;red&quot;</span>, <span class="st">&quot;darkred&quot;</span>)</span>
<span id="cb411-71"><a href="viz-chapter.html#cb411-71" tabindex="-1"></a></span>
<span id="cb411-72"><a href="viz-chapter.html#cb411-72" tabindex="-1"></a>heatmap <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(melted_mat) <span class="sc">+</span> </span>
<span id="cb411-73"><a href="viz-chapter.html#cb411-73" tabindex="-1"></a>  <span class="fu">geom_tile</span>(<span class="fu">aes</span>(<span class="at">x =</span> Sample, <span class="at">y =</span> Taxa, <span class="at">fill =</span> clr_z)) <span class="sc">+</span></span>
<span id="cb411-74"><a href="viz-chapter.html#cb411-74" tabindex="-1"></a>  <span class="fu">theme</span>(</span>
<span id="cb411-75"><a href="viz-chapter.html#cb411-75" tabindex="-1"></a>    <span class="at">axis.title.y=</span><span class="fu">element_blank</span>(),</span>
<span id="cb411-76"><a href="viz-chapter.html#cb411-76" tabindex="-1"></a>    <span class="at">axis.title.x=</span><span class="fu">element_blank</span>(),</span>
<span id="cb411-77"><a href="viz-chapter.html#cb411-77" tabindex="-1"></a>    <span class="at">axis.ticks.y=</span><span class="fu">element_blank</span>(),</span>
<span id="cb411-78"><a href="viz-chapter.html#cb411-78" tabindex="-1"></a>    <span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">90</span>, <span class="at">vjust =</span> <span class="fl">0.5</span>, <span class="at">hjust=</span><span class="dv">1</span>),</span>
<span id="cb411-79"><a href="viz-chapter.html#cb411-79" tabindex="-1"></a>    </span>
<span id="cb411-80"><a href="viz-chapter.html#cb411-80" tabindex="-1"></a>    <span class="at">plot.margin=</span><span class="fu">margin</span>(<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>), <span class="co"># removes margins</span></span>
<span id="cb411-81"><a href="viz-chapter.html#cb411-81" tabindex="-1"></a>    <span class="at">legend.key.height=</span> <span class="fu">unit</span>(<span class="dv">1</span>, <span class="st">&#39;cm&#39;</span>)</span>
<span id="cb411-82"><a href="viz-chapter.html#cb411-82" tabindex="-1"></a>    ) <span class="sc">+</span></span>
<span id="cb411-83"><a href="viz-chapter.html#cb411-83" tabindex="-1"></a>  <span class="fu">scale_fill_gradientn</span>(<span class="at">name =</span> <span class="st">&quot;CLR + Z transform&quot;</span>, </span>
<span id="cb411-84"><a href="viz-chapter.html#cb411-84" tabindex="-1"></a>                       <span class="at">breaks =</span> breaks, </span>
<span id="cb411-85"><a href="viz-chapter.html#cb411-85" tabindex="-1"></a>                       <span class="at">limits =</span> limits, </span>
<span id="cb411-86"><a href="viz-chapter.html#cb411-86" tabindex="-1"></a>                       <span class="at">colours =</span> colours) <span class="sc">+</span> </span>
<span id="cb411-87"><a href="viz-chapter.html#cb411-87" tabindex="-1"></a>  <span class="fu">scale_y_discrete</span>(<span class="at">position =</span> <span class="st">&quot;right&quot;</span>)</span>
<span id="cb411-88"><a href="viz-chapter.html#cb411-88" tabindex="-1"></a></span>
<span id="cb411-89"><a href="viz-chapter.html#cb411-89" tabindex="-1"></a>heatmap</span></code></pre></div>
<p><img src="19_visualization_techniques_files/figure-html/more_complex_heatmap-1.png" width="672" /></p>
<div class="sourceCode" id="cb412"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb412-1"><a href="viz-chapter.html#cb412-1" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span>
<span id="cb412-2"><a href="viz-chapter.html#cb412-2" tabindex="-1"></a></span>
<span id="cb412-3"><a href="viz-chapter.html#cb412-3" tabindex="-1"></a><span class="co"># Create layout</span></span>
<span id="cb412-4"><a href="viz-chapter.html#cb412-4" tabindex="-1"></a>design <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb412-5"><a href="viz-chapter.html#cb412-5" tabindex="-1"></a>  patchwork<span class="sc">::</span><span class="fu">area</span>(<span class="dv">3</span>, <span class="dv">1</span>, <span class="dv">4</span>, <span class="dv">1</span>),</span>
<span id="cb412-6"><a href="viz-chapter.html#cb412-6" tabindex="-1"></a>  patchwork<span class="sc">::</span><span class="fu">area</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">1</span>, <span class="dv">3</span>),</span>
<span id="cb412-7"><a href="viz-chapter.html#cb412-7" tabindex="-1"></a>  patchwork<span class="sc">::</span><span class="fu">area</span>(<span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">3</span>),</span>
<span id="cb412-8"><a href="viz-chapter.html#cb412-8" tabindex="-1"></a>  patchwork<span class="sc">::</span><span class="fu">area</span>(<span class="dv">3</span>, <span class="dv">2</span>, <span class="dv">4</span>, <span class="dv">3</span>)</span>
<span id="cb412-9"><a href="viz-chapter.html#cb412-9" tabindex="-1"></a>)</span>
<span id="cb412-10"><a href="viz-chapter.html#cb412-10" tabindex="-1"></a><span class="co"># to view the design, run</span></span>
<span id="cb412-11"><a href="viz-chapter.html#cb412-11" tabindex="-1"></a><span class="co"># plot(design)</span></span>
<span id="cb412-12"><a href="viz-chapter.html#cb412-12" tabindex="-1"></a></span>
<span id="cb412-13"><a href="viz-chapter.html#cb412-13" tabindex="-1"></a><span class="co"># Combine plots</span></span>
<span id="cb412-14"><a href="viz-chapter.html#cb412-14" tabindex="-1"></a>plot <span class="ot">&lt;-</span> row_annotation <span class="sc">+</span> sample_clusters_annotation <span class="sc">+</span></span>
<span id="cb412-15"><a href="viz-chapter.html#cb412-15" tabindex="-1"></a>                         sample_types_annotation <span class="sc">+</span></span>
<span id="cb412-16"><a href="viz-chapter.html#cb412-16" tabindex="-1"></a>             heatmap  <span class="sc">+</span></span>
<span id="cb412-17"><a href="viz-chapter.html#cb412-17" tabindex="-1"></a>    <span class="fu">plot_layout</span>(<span class="at">design =</span> design, <span class="at">guides =</span> <span class="st">&quot;collect&quot;</span>,</span>
<span id="cb412-18"><a href="viz-chapter.html#cb412-18" tabindex="-1"></a>                <span class="co"># Specify layout, collect legends</span></span>
<span id="cb412-19"><a href="viz-chapter.html#cb412-19" tabindex="-1"></a>                </span>
<span id="cb412-20"><a href="viz-chapter.html#cb412-20" tabindex="-1"></a>                <span class="co"># Adjust widths and heights to align plots.</span></span>
<span id="cb412-21"><a href="viz-chapter.html#cb412-21" tabindex="-1"></a>                <span class="co"># When annotation plot is larger, it might not fit into</span></span>
<span id="cb412-22"><a href="viz-chapter.html#cb412-22" tabindex="-1"></a>        <span class="co"># its column/row.</span></span>
<span id="cb412-23"><a href="viz-chapter.html#cb412-23" tabindex="-1"></a>                <span class="co"># Then you need to make column/row larger.</span></span>
<span id="cb412-24"><a href="viz-chapter.html#cb412-24" tabindex="-1"></a>                </span>
<span id="cb412-25"><a href="viz-chapter.html#cb412-25" tabindex="-1"></a>                <span class="co"># Relative widths and heights of each column and row:</span></span>
<span id="cb412-26"><a href="viz-chapter.html#cb412-26" tabindex="-1"></a>                <span class="co"># Currently, the width of the first column is 15 % and the height of</span></span>
<span id="cb412-27"><a href="viz-chapter.html#cb412-27" tabindex="-1"></a>                <span class="co"># first two rows are 30 % the size of others</span></span>
<span id="cb412-28"><a href="viz-chapter.html#cb412-28" tabindex="-1"></a>                </span>
<span id="cb412-29"><a href="viz-chapter.html#cb412-29" tabindex="-1"></a>                <span class="co"># To get this work most of the times, you can adjust all sizes to be 1, i.e. equal, </span></span>
<span id="cb412-30"><a href="viz-chapter.html#cb412-30" tabindex="-1"></a>                <span class="co"># but then the gaps between plots are larger.</span></span>
<span id="cb412-31"><a href="viz-chapter.html#cb412-31" tabindex="-1"></a>                <span class="at">widths =</span> <span class="fu">c</span>(<span class="fl">0.15</span>, <span class="dv">1</span>, <span class="dv">1</span>),</span>
<span id="cb412-32"><a href="viz-chapter.html#cb412-32" tabindex="-1"></a>                <span class="at">heights =</span> <span class="fu">c</span>(<span class="fl">0.3</span>, <span class="fl">0.3</span>, <span class="dv">1</span>, <span class="dv">1</span>))</span>
<span id="cb412-33"><a href="viz-chapter.html#cb412-33" tabindex="-1"></a></span>
<span id="cb412-34"><a href="viz-chapter.html#cb412-34" tabindex="-1"></a><span class="co"># plot</span></span></code></pre></div>
<div class="sourceCode" id="cb413"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb413-1"><a href="viz-chapter.html#cb413-1" tabindex="-1"></a><span class="co"># Create layout</span></span>
<span id="cb413-2"><a href="viz-chapter.html#cb413-2" tabindex="-1"></a>design <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb413-3"><a href="viz-chapter.html#cb413-3" tabindex="-1"></a>  patchwork<span class="sc">::</span><span class="fu">area</span>(<span class="dv">4</span>, <span class="dv">1</span>, <span class="dv">5</span>, <span class="dv">1</span>),</span>
<span id="cb413-4"><a href="viz-chapter.html#cb413-4" tabindex="-1"></a>  patchwork<span class="sc">::</span><span class="fu">area</span>(<span class="dv">4</span>, <span class="dv">2</span>, <span class="dv">5</span>, <span class="dv">2</span>),</span>
<span id="cb413-5"><a href="viz-chapter.html#cb413-5" tabindex="-1"></a>  patchwork<span class="sc">::</span><span class="fu">area</span>(<span class="dv">1</span>, <span class="dv">3</span>, <span class="dv">1</span>, <span class="dv">4</span>),</span>
<span id="cb413-6"><a href="viz-chapter.html#cb413-6" tabindex="-1"></a>  patchwork<span class="sc">::</span><span class="fu">area</span>(<span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">2</span>, <span class="dv">4</span>),</span>
<span id="cb413-7"><a href="viz-chapter.html#cb413-7" tabindex="-1"></a>  patchwork<span class="sc">::</span><span class="fu">area</span>(<span class="dv">3</span>, <span class="dv">3</span>, <span class="dv">3</span>, <span class="dv">4</span>),</span>
<span id="cb413-8"><a href="viz-chapter.html#cb413-8" tabindex="-1"></a>  patchwork<span class="sc">::</span><span class="fu">area</span>(<span class="dv">4</span>, <span class="dv">3</span>, <span class="dv">5</span>, <span class="dv">4</span>)</span>
<span id="cb413-9"><a href="viz-chapter.html#cb413-9" tabindex="-1"></a>)</span>
<span id="cb413-10"><a href="viz-chapter.html#cb413-10" tabindex="-1"></a></span>
<span id="cb413-11"><a href="viz-chapter.html#cb413-11" tabindex="-1"></a><span class="co"># to view the design, run</span></span>
<span id="cb413-12"><a href="viz-chapter.html#cb413-12" tabindex="-1"></a><span class="co"># plot(design)</span></span>
<span id="cb413-13"><a href="viz-chapter.html#cb413-13" tabindex="-1"></a></span>
<span id="cb413-14"><a href="viz-chapter.html#cb413-14" tabindex="-1"></a><span class="co"># Combine plots</span></span>
<span id="cb413-15"><a href="viz-chapter.html#cb413-15" tabindex="-1"></a>plot <span class="ot">&lt;-</span> taxa_tree <span class="sc">+</span> </span>
<span id="cb413-16"><a href="viz-chapter.html#cb413-16" tabindex="-1"></a>  row_annotation <span class="sc">+</span></span>
<span id="cb413-17"><a href="viz-chapter.html#cb413-17" tabindex="-1"></a>  sample_tree <span class="sc">+</span> </span>
<span id="cb413-18"><a href="viz-chapter.html#cb413-18" tabindex="-1"></a>  sample_clusters_annotation <span class="sc">+</span></span>
<span id="cb413-19"><a href="viz-chapter.html#cb413-19" tabindex="-1"></a>  sample_types_annotation <span class="sc">+</span></span>
<span id="cb413-20"><a href="viz-chapter.html#cb413-20" tabindex="-1"></a>  heatmap <span class="sc">+</span></span>
<span id="cb413-21"><a href="viz-chapter.html#cb413-21" tabindex="-1"></a>    <span class="fu">plot_layout</span>(<span class="at">design =</span> design, <span class="at">guides =</span> <span class="st">&quot;collect&quot;</span>, <span class="co"># Specify layout, collect legends</span></span>
<span id="cb413-22"><a href="viz-chapter.html#cb413-22" tabindex="-1"></a>                <span class="at">widths =</span> <span class="fu">c</span>(<span class="fl">0.2</span>, <span class="fl">0.15</span>, <span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">1</span>),</span>
<span id="cb413-23"><a href="viz-chapter.html#cb413-23" tabindex="-1"></a>                <span class="at">heights =</span> <span class="fu">c</span>(<span class="fl">0.1</span>, <span class="fl">0.15</span>, <span class="fl">0.15</span>, <span class="fl">0.25</span>, <span class="dv">1</span>, <span class="dv">1</span>))</span>
<span id="cb413-24"><a href="viz-chapter.html#cb413-24" tabindex="-1"></a></span>
<span id="cb413-25"><a href="viz-chapter.html#cb413-25" tabindex="-1"></a>plot</span></code></pre></div>
<p>Heatmaps find several other applications in biclustering and
multi-assay analyses. These are discussed further in chapters
<a href="clustering.html#clustering">10</a> and <a href="multi-assay-analyses.html#multi-assay-analyses">13</a>.</p>

</div>
</div>
</div>



            </section>

          </div>
        </div>
      </div>
<a href="multi-assay-analyses.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="training.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": true,
"facebook": false,
"twitter": true,
"linkedin": true,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/microbiome/OMA/blob/master/19_visualization_techniques.Rmd",
"text": "Edit"
},
"history": {
"link": "https://github.com/microbiome/OMA/commits/master/19_visualization_techniques.Rmd",
"text": null
},
"view": {
"link": "https://github.com/microbiome/OMA/blob/master/19_visualization_techniques.Rmd",
"text": null
},
"download": ["OMA.pdf"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
